{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle as pickle\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define snakemake variables\n",
    "#Change to code cell before running snakemake\n",
    "\n",
    "clean_primary_data_path = str(snakemake.input[0])\n",
    "sampled_genes_path = str(snakemake.input[1])\n",
    "divergence_path = str(snakemake.input[2])\n",
    "\n",
    "condition_sets= snakemake.config['condition_sets']\n",
    "\n",
    "type_of_study = snakemake.config['type_of_study']\n",
    "not_type_of_study = snakemake.config['not_type_of_study']\n",
    "pathway = snakemake.config['pathway']\n",
    "\n",
    "MANIPULATED_GENE = snakemake.config['MANIPULATED_GENE']\n",
    "MIN_NUM_SAMPLED = snakemake.config['MIN_NUM_SAMPLED']\n",
    "MIN_NUM_DETECTED = snakemake.config['MIN_NUM_DETECTED']\n",
    "\n",
    "NC_POOL_SOURCE = snakemake.config['NC_POOL_SOURCE']\n",
    "\n",
    "output_path = str(snakemake.params.prefix)\n",
    "\n",
    "local_params = [type_of_study, not_type_of_study,MANIPULATED_GENE,MIN_NUM_SAMPLED, MIN_NUM_DETECTED,pathway]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Define notebook variables\n",
    "#Change to markdown before running snakemake\n",
    "clean_primary_data_path = \"../../data/workflow/clean_primary_data.pk\"\n",
    "sampled_genes_path = \"../../data/workflow/sampled_genes.pk\"\n",
    "divergence_path = \"../../data/workflow/SFS_and_divergence.pk\"\n",
    "\n",
    "condition_sets= [1,2]\n",
    "type_of_study = ['transcriptomics']\n",
    "not_type_of_study = ['RT_qPCR']\n",
    "MANIPULATED_GENE = 2\n",
    "MIN_NUM_SAMPLED = 0\n",
    "MIN_NUM_DETECTED = 0\n",
    "pathway = []\n",
    "NC_POOL_SOURCE = 'genome'\n",
    "local_params = [type_of_study, not_type_of_study,MANIPULATED_GENE,MIN_NUM_SAMPLED, MIN_NUM_DETECTED,pathway]\n",
    "candidate_output_path = \"../../data/workflow/candidate_genes.pk\"\n",
    "noncandidate_output_path = \"../../data/workflow/noncandidate_genes.pk\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_dictionary(snakemake_config_var, condition_sets):\n",
    "    \n",
    "    '''\n",
    "    This function checks whether filter conditions from snakemake config parameters\n",
    "    are in dictionary format. \n",
    "    - If the conditions are not in dictionary format, it is converted to dictionary format.\n",
    "    - When there are multiple condition sets but the condition is not in dictionary format, \n",
    "    assume the condition applies to all condition sets.\n",
    "    '''\n",
    "    var_as_Dict = {}\n",
    "    try:\n",
    "        items = snakemake_config_var.items()\n",
    "    except (AttributeError, TypeError):\n",
    "        for i in condition_sets:\n",
    "            var_as_Dict[i] = snakemake_config_var\n",
    "    return var_as_Dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_by_study(clean_primary_data_path = clean_primary_data_path, \n",
    "                    MANIPULATED_GENE = MANIPULATED_GENE, \n",
    "                    type_of_study = type_of_study, \n",
    "                    not_type_of_study = not_type_of_study):\n",
    "    \n",
    "    '''\n",
    "    This function filters which studies to include in candidate and noncandidate gene pools\n",
    "    MANIPULATED_GENE = <int>\n",
    "    type_of_study = <list>\n",
    "    not_type_of_study = <list>\n",
    "    '''\n",
    "    df = pd.read_pickle(clean_primary_data_path)\n",
    "    \n",
    "    # Did the study manipulated a gene or not?\n",
    "    try:\n",
    "        if MANIPULATED_GENE == 2:\n",
    "            condition_1 = (df.source != None)\n",
    "        if MANIPULATED_GENE == 1:\n",
    "            condition_1 = (df.manipulated_gene == 1)\n",
    "        if MANIPULATED_GENE == 0:\n",
    "            condition_1 = (df.manipulated_gene == 0)\n",
    "    except NameError:\n",
    "        condition_1 = (df.source != None)\n",
    "    \n",
    "    # Is the study included in type_of_study or not_type_of_study?\n",
    "    try:\n",
    "        if len(type_of_study) != 0:\n",
    "            # df['match'] is False if study type is not in type_of_study and is in not_type_of_study\n",
    "            df['match'] = df.apply(lambda x:\n",
    "                               True if (any(item in x.type_of_study for item in type_of_study))&(~any(item in x.type_of_study for item in not_type_of_study)) \\\n",
    "                               else False, \n",
    "                               axis = 1)\n",
    "        # df['match'] is False if study type is in not_type_of_study\n",
    "        else: \n",
    "            df['match'] = df.apply(lambda x:\n",
    "                               True if ~any(item in x.type_of_study for item in not_type_of_study) \\\n",
    "                               else False, \n",
    "                               axis = 1)\n",
    "    except NameError:\n",
    "        df['match'] = True\n",
    "    \n",
    "    #Include studies that are matches, otherwise include every study that has valid name\n",
    "    try:\n",
    "        condition_2 = (df.match == True)\n",
    "    except AttributeError: \n",
    "        condition_2 = (df.source != None)\n",
    "        \n",
    "    selected_studies = list(df[condition_1 & condition_2].source)\n",
    "    \n",
    "    return selected_studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_studies = []\n",
    "def filter_by_columns(selected_studies = selected_studies,\n",
    "                      sampled_genes_path = sampled_genes_path,\n",
    "                      MIN_NUM_SAMPLED = MIN_NUM_SAMPLED, \n",
    "                      MIN_NUM_DETECTED = MIN_NUM_DETECTED):\n",
    "\n",
    "    '''\n",
    "    This function returns a list of PAC_ids of the candidate and \n",
    "    noncandidate (from sampled genes) genes\n",
    "    '''\n",
    "    \n",
    "    #Subset sampled genes by filtered studies\n",
    "    df = pd.read_pickle(sampled_genes_path)\n",
    "    df = df[df.source.isin(selected_studies)]\n",
    "    \n",
    "    #Sum counts in each column for each gene\n",
    "    df = df[['num_detected', 'num_sampled','transcript_id_v5.3.1','pathway_id','PAC_id']]\n",
    "    df = df.groupby(['transcript_id_v5.3.1','pathway_id','PAC_id']).agg('sum').reset_index()\n",
    "        \n",
    "    #Filter by column conditions\n",
    "    if (MIN_NUM_SAMPLED != 0) & (MIN_NUM_DETECTED != 0):\n",
    "        candidate = list(df[(df.num_sampled >= MIN_NUM_SAMPLED) & \\\n",
    "              (df.num_detected >= MIN_NUM_DETECTED)].PAC_id)\n",
    "        noncandidate = list(df.loc[(df.num_sampled >= MIN_NUM_SAMPLED) & \\\n",
    "              (df.num_detected < MIN_NUM_DETECTED)].PAC_id)\n",
    "    else: \n",
    "        candidate = list(df.PAC_id)\n",
    "        noncandidate = list(df.PAC_id)\n",
    "    return candidate, noncandidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_by_pathway(sampled_genes_path = sampled_genes_path, \n",
    "                      pathway = pathway):\n",
    "    \n",
    "    '''\n",
    "    This function returns a list of candidate and noncandidate \n",
    "    (from all genes) genes based on pathway information\n",
    "    '''\n",
    "    \n",
    "    df = pd.read_pickle(sampled_genes_path)\n",
    "    \n",
    "    try:\n",
    "        if len(pathway)!= 0:\n",
    "            df['match'] = df.apply(lambda x: any(item in x.pathway_id for item in pathway), axis = 1)\n",
    "            condition = (df.match == True)\n",
    "    except NameError:\n",
    "        raise NameError\n",
    "    \n",
    "    try: \n",
    "        df.match\n",
    "        candidate = list(df[condition].PAC_id)\n",
    "        noncandidate = list(df.loc[~condition].PAC_id)\n",
    "    except AttributeError: \n",
    "        condition = (df.PAC_id != None)\n",
    "        candidate = list(df[condition].PAC_id)\n",
    "        noncandidate = list(df.loc[condition].PAC_id)\n",
    "    \n",
    "    return candidate, noncandidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_candidate_pool(clean_primary_data_path = clean_primary_data_path,\n",
    "                       sampled_genes_path = sampled_genes_path,\n",
    "                       divergence_path = divergence_path, \n",
    "                       MANIPULATED_GENE = MANIPULATED_GENE, \n",
    "                       type_of_study = type_of_study, \n",
    "                       not_type_of_study = not_type_of_study, \n",
    "                       pathway = pathway, \n",
    "                       MIN_NUM_SAMPLED = MIN_NUM_SAMPLED, \n",
    "                       MIN_NUM_DETECTED = MIN_NUM_DETECTED):\n",
    "    \n",
    "    '''\n",
    "    This function returns the list of candidate and noncandidate genes \n",
    "    based on one set of candidate gene conditions\n",
    "    '''\n",
    "    \n",
    "    #Get list of studies to include\n",
    "    selected_studies = filter_by_study(clean_primary_data_path = clean_primary_data_path, \n",
    "                                       MANIPULATED_GENE = MANIPULATED_GENE, \n",
    "                                       type_of_study = type_of_study, \n",
    "                                       not_type_of_study = not_type_of_study)\n",
    "        \n",
    "    #Filter candidate and noncandidate genes by columns: 'num_detected', \n",
    "                                                        #'num_sampled',\n",
    "                                                        #'transcript_id_v5.3.1',\n",
    "                                                        #'pathway_id',\n",
    "                                                        #'PAC_id'\n",
    "    \n",
    "    candidate_col, noncandidate_col = filter_by_columns(selected_studies = selected_studies,\n",
    "                                                        sampled_genes_path = sampled_genes_path,\n",
    "                                                        MIN_NUM_SAMPLED = MIN_NUM_SAMPLED, \n",
    "                                                        MIN_NUM_DETECTED = MIN_NUM_DETECTED)\n",
    "    \n",
    "    #Filter candidate and noncandidate genes by pathway\n",
    "    candidate_path, noncandidate_path = filter_by_pathway(sampled_genes_path = sampled_genes_path,\n",
    "                                                          pathway = pathway)\n",
    "    \n",
    "    #Get intersection of candidate_col and candidate_path genes\n",
    "    candidate = list(set(candidate_col) & set(candidate_path))\n",
    "    noncandidate_tmp = list(set(noncandidate_col) & set(noncandidate_path))\n",
    "    noncandidate = [i for i in noncandidate_tmp if i not in candidate]\n",
    "    return candidate, noncandidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_temp, noncandidate_temp = get_candidate_pool(clean_primary_data_path = clean_primary_data_path,\n",
    "                                        sampled_genes_path = sampled_genes_path,\n",
    "                                        divergence_path = divergence_path, \n",
    "                                        MANIPULATED_GENE = MANIPULATED_GENE, \n",
    "                                        type_of_study = type_of_study, \n",
    "                                        not_type_of_study = not_type_of_study, \n",
    "                                        pathway = [], \n",
    "                                        MIN_NUM_SAMPLED = MIN_NUM_SAMPLED, \n",
    "                                        MIN_NUM_DETECTED = MIN_NUM_DETECTED)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert all config parameters to dictionary format\n",
    "local_params = [convert_to_dictionary(i, condition_sets) for i in local_params]\n",
    "    \n",
    "candidate = []\n",
    "noncandidate = []\n",
    "for i in condition_sets:\n",
    "    type_of_study, not_type_of_study,MANIPULATED_GENE, MIN_NUM_SAMPLED, MIN_NUM_DETECTED, pathway \\\n",
    "    = [local_params[j][i] for j in range(len(local_params))]\n",
    "        \n",
    "    candidate_temp, noncandidate_temp = get_candidate_pool(clean_primary_data_path = clean_primary_data_path,\n",
    "                                        sampled_genes_path = sampled_genes_path,\n",
    "                                        divergence_path = divergence_path, \n",
    "                                        MANIPULATED_GENE = MANIPULATED_GENE, \n",
    "                                        type_of_study = type_of_study, \n",
    "                                        not_type_of_study = not_type_of_study, \n",
    "                                        pathway = pathway, \n",
    "                                        MIN_NUM_SAMPLED = MIN_NUM_SAMPLED, \n",
    "                                        MIN_NUM_DETECTED = MIN_NUM_DETECTED)\n",
    "    candidate += candidate_temp\n",
    "    noncandidate += noncandidate_temp\n",
    "        \n",
    "#Get SFS of gene pools\n",
    "df = pd.read_pickle(divergence_path)\n",
    "\n",
    "candidate_ls = list(np.unique(candidate))\n",
    "noncandidate_ls = list(np.unique(noncandidate))\n",
    "\n",
    "candidate = df[df.PAC_id.isin(candidate)]\n",
    "\n",
    "#if NC_POOL_SOURCE = \"genome\", noncandidate genes are ALL genes IN THE GENOME that are not candidate genes\n",
    "if NC_POOL_SOURCE == 'genome':\n",
    "    noncandidate = df[~df.PAC_id.isin(candidate_ls)]\n",
    "\n",
    "#if NC_POOL_SOURCE = \"sampled_genes\", noncandidate genes are sampled genes that are not candidate genes \n",
    "#but meets the MIN_NUM_SAMPLED requirement\n",
    "if NC_POOL_SOURCE == \"sampled_genes\":\n",
    "    noncandidate_final = [i for i in noncandidate_ls if i not in candidate_ls]\n",
    "    noncandidate = df[df.PAC_id.isin(noncandidate_final)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"noncandidate\" in output_path:\n",
    "    noncandidate.to_pickle(output_path)\n",
    "else:\n",
    "    candidate.to_pickle(output_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
