{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregate experimental studies data\n",
    "\n",
    "Output file path: `lipid_selection/data/intermediate_data_02/experimental_studies.pickle`\n",
    "\n",
    "Target dataframe output format:\n",
    "\n",
    "| transcript_id | exp_num_detected | exp_num_sampled |exp_num_no_effect | exp_proportion | exp_source                 | annotation_version |\n",
    "|---------------|------------------|-----------------|------------------|----------------|----------------------------|--------------------|\n",
    "| `<string>`    | `<int>`          | `<int>`         | `<float>`        | `<float>`      | `['Firstauthor_year',...]` | `v5.5`             |\n",
    "\n",
    "**`exp_proportion`:** `<float>` \n",
    "> `exp_proportion` = `exp_num_detected`/`num_exp_source`\n",
    "\n",
    "**`num_exp_source`:** `<int>` number of experimental studies\n",
    "\n",
    "Define **experimental studies** as studies that \n",
    "- Have at these one of the following tags:\n",
    "    > ['knock_out','gene_silence','RT_qPCR','gene_knockout','genetic_transformation']\n",
    "- have at least 1 gene manipulated.\n",
    "- are not one of the studies in `match_summary.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import pickle\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"../../data/raw_data/Summary_of_primary_data.xlsx\", 'Sheet1')\n",
    "df = df[['Genus', 'Species','First_author','Publication_year','Type_of_study','Genes_sampled','Genes_manipulated', 'Genes_with_effect',\"Genes_no_effect\"]]\n",
    "match_summary = pd.read_csv(\"../../data/intermediate_data_02/match_summary.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Genus      Species First_author  Publication_year  \\\n",
      "0  Nannochloropsis          NaN        Zheng              2013   \n",
      "1    Chlamydomonas  reinhardtii           Lv              2013   \n",
      "2    Chlamydomonas  reinhardtii           Lv              2013   \n",
      "3      Fistulifera      solaris       Tanaka              2015   \n",
      "4      Fistulifera      solaris        Osada              2017   \n",
      "\n",
      "     Type_of_study Genes_sampled                     Genes_manipulated  \\\n",
      "0  transcriptomics           NaN                                   NaN   \n",
      "1  transcriptomics           NaN                                   NaN   \n",
      "2        knock_out           NaN  Cre05.g248150.t1; Cre06.g299050.t1.2   \n",
      "3  transcriptomics           NaN                                   NaN   \n",
      "4  transcriptomics           NaN                                   NaN   \n",
      "\n",
      "    Genes_with_effect   Genes_no_effect  \n",
      "0                 NaN               NaN  \n",
      "1                 NaN               NaN  \n",
      "2  Cre06.g299050.t1.2  Cre05.g248150.t1  \n",
      "3                 NaN               NaN  \n",
      "4                 NaN               NaN  \n",
      "  first_author    year  matches  mismatches    total  proportion_of_matches\n",
      "0        Hemme  2014.0   1894.0        86.0   1980.0               0.956566\n",
      "1     Bajhaiya  2016.0  17736.0         1.0  17737.0               0.999944\n",
      "2   Goodenough  2014.0     66.0         2.0     68.0               0.970588\n",
      "3        Blaby  2013.0    591.0        51.0    642.0               0.920561\n",
      "4        Boyle  2012.0     24.0         1.0     25.0               0.960000\n"
     ]
    }
   ],
   "source": [
    "print(df.head())\n",
    "print(match_summary.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the types of studies?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "types_of_studies = []\n",
    "\n",
    "for study in list(df.Type_of_study.unique()):\n",
    "    try: \n",
    "        study = study.split(\"; \")\n",
    "        for i in study: \n",
    "            if not i in types_of_studies: types_of_studies.append(i)\n",
    "    except: pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['transcriptomics', 'knock_out', 'gene_silence', 'proteomics', 'RT_qPCR', 'gene_knockout', 'genomics', 'genetic_transformation']\n"
     ]
    }
   ],
   "source": [
    "print(types_of_studies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset data to experimental studies only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "re.compile('knock_out|gene_silence|RT_qPCR|gene_knockout|genetic_transformation')\n",
      "re.compile('Hemme|Bajhaiya|Goodenough|Blaby|Boyle|Juergens|Kwak|Gargouri')\n"
     ]
    }
   ],
   "source": [
    "pattern_study = re.compile(\"|\".join(['knock_out','gene_silence','RT_qPCR','gene_knockout','genetic_transformation']))\n",
    "print(pattern_study)\n",
    "\n",
    "pattern_author = re.compile(\"|\".join(list(match_summary.first_author)))\n",
    "print(pattern_author)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Genus      Species First_author  Publication_year Type_of_study  \\\n",
      "2   Chlamydomonas  reinhardtii           Lv              2013     knock_out   \n",
      "9   Chlamydomonas  reinhardtii        Siaut              2011     knock_out   \n",
      "10  Chlamydomonas  reinhardtii         Wang              2009     knock_out   \n",
      "11  Chlamydomonas  reinhardtii   Moellering              2009  gene_silence   \n",
      "16  Chlamydomonas  reinhardtii           Li              2010     knock_out   \n",
      "\n",
      "   Genes_sampled                     Genes_manipulated  \\\n",
      "2            NaN  Cre05.g248150.t1; Cre06.g299050.t1.2   \n",
      "9            NaN                STA1; STA6; STA7; STA8   \n",
      "10           NaN                                  STA6   \n",
      "11           NaN                                  MLDP   \n",
      "16           NaN         STA6; STA7; STA1; STA11; ARG7   \n",
      "\n",
      "                Genes_with_effect   Genes_no_effect  \n",
      "2              Cre06.g299050.t1.2  Cre05.g248150.t1  \n",
      "9          STA1; STA6; STA7; STA8               NaN  \n",
      "10                           STA6               NaN  \n",
      "11                           MLDP               NaN  \n",
      "16  STA6; STA7; STA1; STA11; ARG7               NaN   \n",
      "\n",
      " (23, 9)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "subset = df[df.Type_of_study.str.contains(pattern_study)==True][df.Genes_manipulated.isnull()==False][df.First_author.str.contains(pattern_author)==False]\n",
    "print(subset.head(), '\\n'*2,subset.shape)\n",
    "num_studies = subset.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Genus      Species First_author  Publication_year  \\\n",
      "2   Chlamydomonas  reinhardtii           Lv              2013   \n",
      "9   Chlamydomonas  reinhardtii        Siaut              2011   \n",
      "10  Chlamydomonas  reinhardtii         Wang              2009   \n",
      "11  Chlamydomonas  reinhardtii   Moellering              2009   \n",
      "16  Chlamydomonas  reinhardtii           Li              2010   \n",
      "\n",
      "      Type_of_study Genes_sampled                       Genes_manipulated  \\\n",
      "2      (knock_out,)           NaN  (Cre05.g248150.t1, Cre06.g299050.t1.2)   \n",
      "9      (knock_out,)           NaN                (STA1, STA6, STA7, STA8)   \n",
      "10     (knock_out,)           NaN                                 (STA6,)   \n",
      "11  (gene_silence,)           NaN                                 (MLDP,)   \n",
      "16     (knock_out,)           NaN         (STA6, STA7, STA1, STA11, ARG7)   \n",
      "\n",
      "                  Genes_with_effect      Genes_no_effect  \n",
      "2             (Cre06.g299050.t1.2,)  (Cre05.g248150.t1,)  \n",
      "9          (STA1, STA6, STA7, STA8)                  NaN  \n",
      "10                          (STA6,)                  NaN  \n",
      "11                          (MLDP,)                  NaN  \n",
      "16  (STA6, STA7, STA1, STA11, ARG7)                  NaN  \n"
     ]
    }
   ],
   "source": [
    "for col in df.columns[4:]:\n",
    "    subset[col] = subset[col].apply(lambda x: tuple(x.split(\"; \")) if isinstance(x,str) else np.NaN) \n",
    "print(subset.head())  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Genus', 'Species', 'First_author', 'Publication_year', 'Type_of_study',\n",
      "       'Genes_sampled', 'Genes_manipulated', 'Genes_with_effect',\n",
      "       'Genes_no_effect'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(subset.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reaarrange subset to target datframe format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert 20201019_sampled_genes_key.csv to dictionary format with `symbol` as keys and `transcript_id` as values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                              source symbol  \\\n",
      "0  Sim, L., Beeren, S. R., Findinier, J., Dauvill...   STA8   \n",
      "1                                        Phytozome12   APL1   \n",
      "2                                        Phytozome12  PEPC1   \n",
      "3                                              Query   NRR1   \n",
      "4                                              Query   SQD1   \n",
      "\n",
      "        transcript_id  \n",
      "0  Cre17.g698850.t1.1  \n",
      "1  Cre16.g683450.t1.1  \n",
      "2       Cre16.g673852  \n",
      "3  Cre16.g673250.t1.1  \n",
      "4  Cre16.g656400.t1.2  \n",
      "[('Cre05.g248150.t1', 'Cre05.g248150.t1.2'), ('SCD', 'Cre09.g397250'), ('ARG7', 'Cre01.g021251.t1.1'), ('FEA1', 'Cre12.g546550.t1.1'), ('ACS2', 'Cre01.g055408.t1.1')]\n"
     ]
    }
   ],
   "source": [
    "sampled_genes_key = pd.read_csv(\"../../analysis/01_generate_intermediate_data_01/20201019_sampled_genes_key.csv\")\n",
    "print(sampled_genes_key.head())\n",
    "#print()\n",
    "key = dict(zip(sampled_genes_key.symbol, sampled_genes_key.transcript_id))\n",
    "print(list(key.items())[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset = pd.DataFrame(columns = ['transcript_id','exp_num_sampled', 'exp_num_manipulated', 'exp_num_detected', 'exp_num_no_effect', 'exp_source'])\n",
    "for index, row in subset.iterrows():\n",
    "    source = row.First_author + \"_\" + str(row.Publication_year)\n",
    "    for i in range(0,4): \n",
    "        if type(row[i+5]) != float:\n",
    "            for gene in row[i+5]: #Genes_sampled\n",
    "                output = [0]*(i)+[1]+[0]*(3-i)\n",
    "                try: \n",
    "                    gene = key[re.sub(r\" |;\",\"\", gene)]\n",
    "                except: gene = re.sub(r\" |;\",\"\", gene)\n",
    "                output = pd.Series([gene]+output+[source], index = transformed_subset.columns)\n",
    "                transformed_subset = transformed_subset.append(output, ignore_index = True)\n",
    "                #print(output)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exp_num_sampled</th>\n",
       "      <th>exp_num_manipulated</th>\n",
       "      <th>exp_num_detected</th>\n",
       "      <th>exp_num_no_effect</th>\n",
       "      <th>exp_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cre05.g248150.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre06.g299050.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre06.g299050.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cre05.g248150.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cre13.g567950.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Siaut_2011</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  exp_num_sampled  exp_num_manipulated  exp_num_detected  \\\n",
       "0  Cre05.g248150.t1.2              0.0                  1.0               0.0   \n",
       "1  Cre06.g299050.t1.2              0.0                  1.0               0.0   \n",
       "2  Cre06.g299050.t1.2              0.0                  0.0               1.0   \n",
       "3  Cre05.g248150.t1.2              0.0                  0.0               0.0   \n",
       "4  Cre13.g567950.t1.2              0.0                  1.0               0.0   \n",
       "\n",
       "   exp_num_no_effect  exp_source  \n",
       "0                0.0     Lv_2013  \n",
       "1                0.0     Lv_2013  \n",
       "2                0.0     Lv_2013  \n",
       "3                1.0     Lv_2013  \n",
       "4                0.0  Siaut_2011  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = transformed_subset.groupby(['transcript_id'])[['exp_num_sampled', 'exp_num_manipulated', 'exp_num_detected', 'exp_num_no_effect']].sum()\n",
    "source = transformed_subset.groupby(['transcript_id'])[['exp_source']].agg(lambda x: list(np.unique(list(itertools.chain(x)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_num_sampled</th>\n",
       "      <th>exp_num_manipulated</th>\n",
       "      <th>exp_num_detected</th>\n",
       "      <th>exp_num_no_effect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp_num_sampled  exp_num_manipulated  exp_num_detected  \\\n",
       "transcript_id                                                                \n",
       "Cre01.g021251.t1.1              0.0                  1.0               1.0   \n",
       "Cre01.g038550.t1.1              1.0                  1.0               1.0   \n",
       "Cre01.g045903.t1.1              3.0                  0.0               1.0   \n",
       "Cre01.g053000.t1.2              1.0                  0.0               1.0   \n",
       "Cre01.g053150.t1.2              1.0                  0.0               1.0   \n",
       "\n",
       "                    exp_num_no_effect  \n",
       "transcript_id                          \n",
       "Cre01.g021251.t1.1                0.0  \n",
       "Cre01.g038550.t1.1                0.0  \n",
       "Cre01.g045903.t1.1                1.0  \n",
       "Cre01.g053000.t1.2                0.0  \n",
       "Cre01.g053150.t1.2                0.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>[Li_2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>[Iwai_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             exp_source\n",
       "transcript_id                                          \n",
       "Cre01.g021251.t1.1                            [Li_2010]\n",
       "Cre01.g038550.t1.1                          [Iwai_2014]\n",
       "Cre01.g045903.t1.1  [Iwai_2014, Msanne_2012, Sato_2014]\n",
       "Cre01.g053000.t1.2                          [Sato_2014]\n",
       "Cre01.g053150.t1.2                          [Sato_2014]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_num_sampled</th>\n",
       "      <th>exp_num_manipulated</th>\n",
       "      <th>exp_num_detected</th>\n",
       "      <th>exp_num_no_effect</th>\n",
       "      <th>exp_source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp_num_sampled  exp_num_manipulated  exp_num_detected  \\\n",
       "transcript_id                                                                \n",
       "Cre01.g021251.t1.1              0.0                  1.0               1.0   \n",
       "Cre01.g038550.t1.1              1.0                  1.0               1.0   \n",
       "Cre01.g045903.t1.1              3.0                  0.0               1.0   \n",
       "Cre01.g053000.t1.2              1.0                  0.0               1.0   \n",
       "Cre01.g053150.t1.2              1.0                  0.0               1.0   \n",
       "\n",
       "                    exp_num_no_effect                           exp_source  \n",
       "transcript_id                                                               \n",
       "Cre01.g021251.t1.1                0.0                            [Li_2010]  \n",
       "Cre01.g038550.t1.1                0.0                          [Iwai_2014]  \n",
       "Cre01.g045903.t1.1                1.0  [Iwai_2014, Msanne_2012, Sato_2014]  \n",
       "Cre01.g053000.t1.2                0.0                          [Sato_2014]  \n",
       "Cre01.g053150.t1.2                0.0                          [Sato_2014]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset = pd.merge(sample, source, left_index = True, right_index= True)\n",
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset['exp_proportion'] = transformed_subset.exp_num_detected / num_studies\n",
    "transformed_subset['annotation_version'] = 'v5.5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_num_sampled</th>\n",
       "      <th>exp_num_manipulated</th>\n",
       "      <th>exp_num_detected</th>\n",
       "      <th>exp_num_no_effect</th>\n",
       "      <th>exp_source</th>\n",
       "      <th>exp_proportion</th>\n",
       "      <th>annotation_version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp_num_sampled  exp_num_manipulated  exp_num_detected  \\\n",
       "transcript_id                                                                \n",
       "Cre01.g021251.t1.1              0.0                  1.0               1.0   \n",
       "Cre01.g038550.t1.1              1.0                  1.0               1.0   \n",
       "Cre01.g045903.t1.1              3.0                  0.0               1.0   \n",
       "Cre01.g053000.t1.2              1.0                  0.0               1.0   \n",
       "Cre01.g053150.t1.2              1.0                  0.0               1.0   \n",
       "\n",
       "                    exp_num_no_effect                           exp_source  \\\n",
       "transcript_id                                                                \n",
       "Cre01.g021251.t1.1                0.0                            [Li_2010]   \n",
       "Cre01.g038550.t1.1                0.0                          [Iwai_2014]   \n",
       "Cre01.g045903.t1.1                1.0  [Iwai_2014, Msanne_2012, Sato_2014]   \n",
       "Cre01.g053000.t1.2                0.0                          [Sato_2014]   \n",
       "Cre01.g053150.t1.2                0.0                          [Sato_2014]   \n",
       "\n",
       "                    exp_proportion annotation_version  \n",
       "transcript_id                                          \n",
       "Cre01.g021251.t1.1        0.043478               v5.5  \n",
       "Cre01.g038550.t1.1        0.043478               v5.5  \n",
       "Cre01.g045903.t1.1        0.043478               v5.5  \n",
       "Cre01.g053000.t1.2        0.043478               v5.5  \n",
       "Cre01.g053150.t1.2        0.043478               v5.5  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset['transcript_id']= transformed_subset.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset.index = range(len(transformed_subset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_num_sampled</th>\n",
       "      <th>exp_num_manipulated</th>\n",
       "      <th>exp_num_detected</th>\n",
       "      <th>exp_num_no_effect</th>\n",
       "      <th>exp_source</th>\n",
       "      <th>exp_proportion</th>\n",
       "      <th>annotation_version</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g021251.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g038550.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g045903.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g053000.t1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g053150.t1.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   exp_num_sampled  exp_num_manipulated  exp_num_detected  exp_num_no_effect  \\\n",
       "0              0.0                  1.0               1.0                0.0   \n",
       "1              1.0                  1.0               1.0                0.0   \n",
       "2              3.0                  0.0               1.0                1.0   \n",
       "3              1.0                  0.0               1.0                0.0   \n",
       "4              1.0                  0.0               1.0                0.0   \n",
       "\n",
       "                            exp_source  exp_proportion annotation_version  \\\n",
       "0                            [Li_2010]        0.043478               v5.5   \n",
       "1                          [Iwai_2014]        0.043478               v5.5   \n",
       "2  [Iwai_2014, Msanne_2012, Sato_2014]        0.043478               v5.5   \n",
       "3                          [Sato_2014]        0.043478               v5.5   \n",
       "4                          [Sato_2014]        0.043478               v5.5   \n",
       "\n",
       "        transcript_id  \n",
       "0  Cre01.g021251.t1.1  \n",
       "1  Cre01.g038550.t1.1  \n",
       "2  Cre01.g045903.t1.1  \n",
       "3  Cre01.g053000.t1.2  \n",
       "4  Cre01.g053150.t1.2  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export `transformed_subset` as `lipid_selection/data/intermediate_data_02/experimental_studies.pickle`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/intermediate_data_02/experimental_studies.pickle', 'wb') as f:\n",
    "    pickle.dump(transformed_subset, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is the distribution of number of exp_source and exp_num_sampled/manipulated for the genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "fig, ax = plt.subplots()\n",
    "plt.hist(transformed_subset.exp_proportion, bins = 10, histtype = 'stepfilled')\n",
    "ax.set_xlabel('Proportion of datasets that detected effect of a unique gene')\n",
    "ax.set_ylabel('Number of unique genes')\n",
    "plt.savefig('simulations_output/distribution_of_exp_proportion.png', bbox_inches = \"tight\", dpi = 300)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEbCAYAAADTZlM/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xm4HGWZ/vHvHQJkARQCMoNRUHFjicMmICBBRBGRnwbEXWQc9YcKCI6oXINRcBkXEEWU0QE0IpvoKMGwiOagIMoSIIEgqARIZAhhEZKwBJJn/njfDp1Od586Oae6zkndn+vq61RX1/J0dZ9+qt6qel5FBGZmVj+jqg7AzMyq4QRgZlZTTgBmZjXlBGBmVlNOAGZmNeUEYGZWU6UmAEnrS7pe0ixJd0g6JY/fStIfJM2WdJ6k0WXGYWZmqys1AUTEU8BrI2JHYBvgNZL2Ab4NfDUiJgELgY+XGYeZma2u9CagiHgiD66f17cQ2C0ifpnHnwMcWHYcZma2qtITgKRRkm4C7gf6gEeAB5smWQA8v+w4zMxsVaW3vUfECmAHSRsBlwM3l71OMzPrX89OvkbEY5JmAC8GNm16aSLpKGA1klyoyMxsDUSE+pum7KuAJkjaIA+PBfYDbgL+KOmtebL3Apd2XkpU+HiIceM2JiK6PqZOndrvNMPh4TjrFaPjrG+cRZV9BLAFME0SwBjg3Ij4laS5wLmSTgTmAp8qOQ4zM2tRagKIiDnADm3GzwN2L3PdZmbWne8EHgKTJ0+uOoRCHOfQGQkxguMcaiMlzqI0kPaiXksngauM72HGjduapUsfrjAGM7OBkURUfRLYzMyGLycAM7OacgIwM6spJwAzs5pyAjAzqyknADOzmnICMDOrKScAM7OacgIwM6spJwAzs5pyAjAzqyknADOzmnICMDOrKScAM7OacgIwM6spJwAzs5pyAjAzqyknADOzmnICMDOrKScAM7OacgIwM6spJwAzs5pyAjAzqyknADOzmnICMDOrKScAM7OaKjUBSJoo6SpJcyT9WdKn8vipkhZImpUf+5cZh5mZrW50yct/GvhYRNwqaQNglqTL82unRMQpJa/fzMw6KDUBRMRCYGEeXiJpNvD8/LLKXLeZmXXXs3MAkrYCdgauzqM+KmmupB9L2rhXcZiZWdKTBJCbf34KHB0Ri4HvAFtHxDbAXcBpvYjDzMyeVfY5ACSNBi4CfhIRvwSIiIeaJjkDmNl5CZ9vGp6cH2Zm1tDX10dfX9+A51NEDH00zSuQpgEPRsSxTeM2i4hFefhIYJ+ImNJm3oBy4+vuYcaN25qlSx+uMAYzs4GRRET0e5611CMASXsA7wHmSLqJ9Gt+PPAeSZOAdYF7gQ+WGYeZma2u9COAwfARgJnZwBU9AvCdwGZmNeUEYGZWU04AZmY15QRgZlZTTgBmZjXlBGBmVlNOAGZmNeUEYGZWU04AZmY15QRgZlZTA0oAkjaWtGNZwZiZWe/0mwAk/U7SeEmbAjcD35P0rfJDMzOzMhU5AtggIpYCU4CzImJXYJ9ywzIzs7IVSQDrStoMOBiYkccN3xKiZmZWSJEE8GWgD7grIq6XtCWpG0czMxvB3B9AV+4PwMxGniHrD0DSNpKulnR70/OpQxGkmZlVp0gT0FnAscCTABExF3h7mUGZmVn5iiSAMRFxXcu45WUEY2ZmvVMkATwi6SXkxnhJBwJuFDczG+FGF5jmI8CPgFdKugdYBLyz1KjMzKx0/SaAiLgT2FPSBNJVQw+WH5aZmZWt3wQgaRzppO8LgFFSurIoIk4sNzQzMytTkSagXwEPADfik79mZmuNIglgs4hw7R8zs7VMkauArpa0bemRmJlZTxU5Angt8K+S5gFPAQIiIiaVGpmZmZWqSAJ4U+lRmJlZz/XbBBQR9wBbA/vl4aVF5jMzs+GtSDG4LwOfAI5rmue8IguXNFHSVZLmSPqzpOPy+I0lXSHpFkmXSXrOmr4BMzNbM0X25N8KHETa8yciHgDGFFz+08DHImJ7YGfgg5ImAV8AZkTEq4DLAN9TYGbWY0USwDOROg1o1AIaA6xXZOERsTAibs3DS4A5wETgzcCP82Tn5OdmZtZDRRLATyX9F/BcSYcDVwA/HOiKJG1FOgr4PenegocAcmmJzQa6PDMzG5witYBOkvQWYBmwE/D1iJg+kJVI2gD4KXB0RCxOPX0V9fmm4cn5YWZmDX19ffT19Q14vtK7hJQ0GrgEuCwiTs3j/grsGhEPSdoUuDYiXtpmXncJaWY2QEPZJeRiSY+1PBZKmiHpZQViOQuY2/jxz2YA78vD7wMuLbAcMzMbQv0eAUg6CbgHuIB0F/DbgZcBs4BjImK3LvPuAfyOdPK3cSL5eOC6vLzNgfuBQyPiH23m9xGAmdkAFT0CKJIAboiInVvGXR8Ru0iaGxHbDDLWbut2AjAzG6AhawIC1pF0sKRR+XEwz548XjaoKM3MrDJFEsA7Sd1CPgr8Iw+/O98PcGSJsZmZWYlKvwpoMNwEZGY2cEPZBGRmZmshJwAzs5pyAjAzq6kiN4JtIekcSVfk5y+X9OHyQzMzszIVOQI4B7iYdNMWwF/x1T9mZiNekQQwISIuBFYARMRy4JlSozIzs9IVSQCPS5rAs/0B7IBvADMzG/GKdAp/LHA58BJJVwEvJNUDMjOzEaxIfwDXStodmEQqBjc7InwEYGY2wvWbACS9v2XUNvkus2klxWRmZj1QpAlol6bhMcC+pFLQTgBmZiNYkSagVS75lLQhqZa/mZmNYGtyJ/CTwMuHOhAzM+utIucApvNsSc5RwLbAL8sMyszMylfkHMA3moZXAAsiYl5J8ZiZWY8UOQdwVS8CMTOz3irSBLSY9r2yCIiI2GjIozIzs9IVaQI6FZgPnEf60X8HMDEippYZmJmZlavfLiEl3RARO/c3rgzuEtLMbOCGskvIUZLeKWkdSaMkvbPgfGZmNowV+SF/B/AB4JH8OCyPMzOzEazIVUB/AfbvQSxmZtZDHROApOMi4muSTqNNQ3xEHFVqZGZmVqpuRwC357839CIQMzPrrX6vAqqSrwIyMxu4IbsKSNK2ks6WdKWk3zYeBYM4U9JCSbObxk2VtEDSrPzw+QUzswoUuRHsZ8C3gO8Bywe4/LOB01i974BTIuKUAS7LzMyGUJEE8FhEfG9NFh4RV0vass1L/R6amJlZuYrcBzBD0kck/bOkTRqPQa73o5LmSvqxpI0HuSwzM1sDRUpBtCv9HBHx4kIrSEcA0yNiUn4+AXg4IkLSF4CXRMR7O8zrk8BmZgNU9CRwkRvBXjQ0Ia1c3kNNT88AZnaf4/NNw5Pzw8zMGvr6+ujr6xvwfEWOAN7fbnxEFOoUXtJWpCOA7fPzzSJiUR4+EtgnIqZ0mNdHAGZmAzRkRwDALk3DY4B9gVmsfmVPuyDOJe2yT5B0LzAVeJ2kScC6wL3ABwvEYGZmQ2zAN4JJ2hC4ICIOKCekVdblIwAzswEaynLQrZ4EXr4G85mZ2TBSpEvI6Ty7Gz4K2Bb4ZZlBmZlZ+YqcA/hG0/AKYEFEtLs01MzMRpAil4Fe1YtAzMyst9y1o5lZTTkBmJnVVMcEIOk3+e9XexeOmZn1SrdzAC+Q9BrgIEnn01LBMyJmlRqZmZmVqlsC+A/gBGAi0Fq7P4DXlRWUmZmVr0gtoBMi4qQexdO6bt8JbGY2QENZDfQkSYcCe+VRV0XERYMN0MzMqlXkTuBvAtsB5+VRH5K0R0QcU2pkZmZWqiJ3Ar8B2C5yW5GkHwK3lhmUmZmVr8h9AAI2anq+Ie7T18xsxCtyBPB14DZJV5J++PcBPldqVGZmVrpC/QFIegGwW356bUQsKDWqZ9frq4DMzAZoKHsEIyLmA/MHHZWZmQ0brgVkZlZTTgBmZjXVNQFIGiXptl4FY2ZmvdM1AUTECuBOSVv0KB4zM+uRIieBJwB3SLoOWNoYGREHlRaVmZmVrkgCOKH0KMzMrOcK9QksaWvgJRFxuaSxReYzM7Phrd+rgCQdCVwAfDeP2hyYXmZQZmZWviKXgR4B7A48BhARdwMblxiTmZn1QJEEsCwiljWeSBoFrFdeSGZm1gtFEsDvJR0PjJW0D3AuMKPcsMzMrGxFEsAngMXAX4BjgWuATxVZuKQzJS2UNLtp3MaSrpB0i6TLJD1nTQI3M7PBKVoNdAypV7AAbo2IpwotXNoTWAJMi4hJedy3gbsi4lRJnwBeFBFHd5jf1UDNzAaoaDXQIlcBvQ24B/gGcApwt6S3FgkiIq4GHmkZ/Wbgx3n4nPzczMx6rGiHMLvmq3+QtBVwJfCLNVznphHxEEBEPChpszVcjpmZDUKRcwAPNn78YeVloIvKCsjMzHqj4xGApCl58CZJlwAXkRrkDwZuGsQ6F0maEBEPSdoUeKD75J9vGp6cH2Zm1tDX10dfX9+A5+t4EljS2d1mjIjDC60gNRlNj4jt8/Pmk8DHkE4CH9VhXp8ENjMboKIngQtdBTSIIM4l7bJPABYCU0nnDi4klZS4Hzg0Iv7RYX4nADOzARqyBJALwR0NvICmcwa9KAftBGBmNnBD2Sn8JaRCcD8HVgw2MDMzGx6KJIBHIuLbpUdiZmY9VSQBnC7pBNK1/yvvAI6IWaVFZWZmpSuSALYD3ge8nmebgAJ4XVlBmZlZ+YokgINJl2ou63dKMzMbMYrcCTwb2LDsQMzMrLeKHAFMAP4i6XpWPQdQ+mWgZmZWniIJYGrpUZiZWc/1mwAi4qpeBGJmZr3VbwKQtJhnb8ddD1gXWBoRG5UZmJmZlavIEcDKE8CSBBwIvKbMoMzMrHxFrgJaKZLpwBtLisfMzHqkSBPQlKano4CdgX6LDJmZ2fBW5CqgtzQNrwAWAAeUE46ZmfVKkXMAhTp+MTOzkaVbl5Cf6zJfRMRJJcRjZmY90u0IYGmbceOBD5LuDnYCMDMbwTomgIg4uTEsaUNSr2CHA+cDJ3eaz8zMRoau5wAkbQIcC7wH+BGwY0Q80ovAzMysXN3OAXwdmAJ8H9g+Ipb0LCozMytdx07hJa0gVf98hlV7ZhfpJHDppSDcKbyZ2cANulP4iBjQXcJmZjay+EfezKymnADMzGrKCcDMrKacAMzMasoJwMysppwAzMxqqkg56FJIuht4lFRi+umIeHVVsZiZ1VFlCYD0wz/ZpSXMzKpRZROQKl6/mVmtVfkDvAK4QtItkj5eYRxmZrVUZRPQ7hHxgKTNgMsk3R4Rv6kwHjOzWqksAUTEA/nvIkkXAbsAbRLA55uGJ+eHmZk19PX10dfXN+D5OlYDLZOkcaSKok9IGg/MAE6OiItbpnM1UDOzARp0NdCSbQ78IpecHgec3/rjb2Zm5aokAUTEPOBVVazbzMwSX4ZpZlZTTgBmZjXlBGBmVlNOAGZmNeUEYGZWU04AZmY15QRgZlZTTgBmZjXlBGBmVlNOAGZmNVVlOWgraObM3/Hd755ddRjssssOHHfcUVWHYWZDpJJqoEW5Gmhy2GFHMG3aEuB1FUbxMM997jd55JEFFcZgZkUM92qgNmCvAQ6vcP0LgG9WuH4zG2o+B2BmVlNOAGZmNeUEYGZWU04AZmY15ZPAVtjy5cuYPXt2pTGMGjWKbbfdFqnfCxxKddddd7FkyZJKY1hnnXXYZpttKt8WNnI5AVhBGxOxLXvt9d5Ko3jyyfmcd96ZTJkypbIY5s+fzytesR1jx25dWQwATz55DxdffCFvfOMbK43DRi4nACtoPEuWzKw6CMaPP4zFixdXGsMTTzzBmDETeeyxao+GNtrokMq3hY1sPgdgZlZTTgBmZjXlBGBmVlNOAGZmNeUEYGZWU74KyMxsCCxatIhDDjmcJUueqDSO8ePHFJ7WCcDMbAjceeed3HjjPJYu/XalcYwff2zhaZ0AzMyGyOjRzwX2rTiGjQtPW9k5AEn7S5oj6TZJn64qDjOzuqokAUhaD/ge8EbgVcAhkv6liliGQl9fX9UhFNRXdQAF9VUdQL+eeebxqkMoZKR8Nx1nNao6AtgVuDUi7ouIZ4ALgDdXFMugjZwvRV/VARTUV3UA/Vq+vNoTfUWNlO+m46xGVQlgIjC/6fmCPM7MzHpkBJwEnl7huh/m8ccf4cMf/nDXqW688Ubuu+++0qKYNu0HwJsYfI68g2q3Z1Gd43ziiemceOLVXHPNNb0Nqcnf//53li17mKq35eLF05k69XauuOKKjtOU/d0cKmtDnHPnzuXRR/9A1d+Lp59+sPC0iogSQ+mwUmkv4NMRcWB+/u/A+hHxpZbpeh+cmdlaICL67SiiqiOA64BtJW0BLALeAXykdaIib8DMzNZMJQkgIp6SdARwBSDgxxExq4pYzMzqqpImIDMzq96wLAYn6UxJCyVV2+VSF5ImSroq38z2Z0nHVR1TO5LWl3S9pFmS7pB0StUxdSNpVI714qpj6UTS3ZJukXSTpOuqjqcTSc+RdGGOda6k3aqOqZWkl+XtOCv/fVTSUVXH1UrSFyTdKel2ST+VNLbqmNqR9Jkc55wi23FYHgFI2hNYAkyLiElVx9OOpM2BzSLiVkkbALOAQyJi2CUtSWMj4glJ6wDXAJ+JiL6Kw2pL0jHATsBGEXFQ1fG0I+kuYKeIeKTqWLqRdCHw84g4X9IoYHxEDNs+JHOMC4BdI2J+f9P3iqSXAL8GXhERyyRdAFweEWdVHNoqJO0I/AjYGXgGuAw4JiJu7TTPsDwCiIirgWH9zxURCxsbNiKWALOB51cbVXsR0bhraX3SZ/5AheF0JGkicADw31XH0g8xTP93GiRtAvxLRJwPEBErhvOPf/Z64G/D6cc/exhYBoyXNBoYB9xbbUhtvQL4Y0Q8FRHLgauAt3SbYVh/iUcKSVuRsu7V1UbSXm5WuQm4H+iLiLlVx9TBN4FPAcPvsHRVK4ArctPKx6sOpoOXAg/mJqBbJf1I0viqg+rHO4Dzqg6iVT7SO5n0o/934B8RcWW1UbU1B9hb0saSxpF2pl7YbQYngEHKzT8/BY4erntYee9vB9KdZK+VtHfVMbWS9GZgYUTcTNrDHs6XAO8eETuR9lgPl1Rt+cf2RgG7AF+LiO1IR9QnVBtSZ5LWBQ4i/S8NK5JeDBwDbAlsAWwg6d3VRrW6iJgDnELa8/8tqVWi686UE8Ag5MPBi4CfRMQvq46nPxHxGPArYNidDAT2AA7K7evnAftImlZxTG1FxAP57yLS579LtRG1NR9YEBE35OcXAcO54OKbgBvzNh1uXg1cExEP56aVnwN7VhxTWxFxRkRMiojdgP8F/txt+uGcAIb7XiDAWcDciDi16kA6kTQhH6WQr1zYD+h4UqgqEXF8RLwwIl4MvBP4bUS8v+q4Wkka17gCJDep7A8Muya1iFhAagJ6aR61L/38GFTsXQzD5p/sb8BuksZKEmlb/q3imNqSNCH//SfgUFKhzY6GZS0gSecCk4EJku4FpkbE2dVGtSpJewDvAebk9vUAjo+Iy6qNbDVbANPS95YxwLkR8atqQxrRNgd+IWkF6WTg+RExXC9Z/Tfg3Jyw7iV9X4ed3F79eqB70a2KRMT1ki4iNaksB24GTq82qo5+IWlD4GngYxGxsNvEw/IyUDMzK99wbgIyM7MSOQGYmdWUE4CZWU05AZiZ1ZQTgJlZTTkBmJnVlBPAWkTSCklfb3r+SUmfG6Jlny1pylAsq5/1HJrL2f6m7HUViOULkl5X8joOyzft9Ddd4e0vaW9Ju69BLCvXIekHkl4x0GV0i2Og36FczvqIwcRg3TkBrF2eAqbkSpDDRi7zW9ThwAciotL6OpJGRcTUiPhtyav6AENfRXYy8JrBLCAiPhQRg71zeLBxbAx8dJAxWBdOAGuXZ4DvA8e2vtC69yVpcf67t6Q+ST+T9BdJ/ynpvZKuVeroZuumxewn6Y+S/irpbXn+dSSdJulmpU5Hjmxa7u8k/Q+pSmFrPIfn6W+T9M087gRSjZUzJX21zTwnSJqd5/tyHrdzrsq5nqTxufLlNnn9V0m6WKkjnLOalvMWSTfmZf2iqVTGvPz+/wgc3LJHPE/SlyTdoNTBzo6SZuTxR/YT45b5+Rk5vpm5pMTBpCqy5yh1iLK+pKmS/qTU8cjZufRAR5KOydvwJknnS9oS+P/AJ/Iy9+jy2Svv6f9Z0qXA85qmmalUX77b9vp6fj83STq5Ja7V4sgvNb4X90h6V552fF7fDfl9H5Kn/Qrw4jz/at8HGwIR4cda8gAeAzYA5gEbAp8EPpdfOxuY0jxt/rs3qd75BGA9UrnbE/JrRwGnNc1/SR7eClhIKi1xJKkEBnn+G4Gt83IXA1u0ifMFpI4/nkOq9/Rr4ND82kxghzbzvAX4rzwsYDqwb35+IvB14DvAcU3v63FgYn5+GanG0POAPwBj8/jjgJPy8DxSBxq0brP82r/l4VOAW0j9K2wKLOoWI6mK5DJg2/zaBcBh7d4vsGHT8LSm9a/y+TVN83dg3Tw8Pv+dChzb7n20fPbvAmbk4eeRKoZOaYprx07bC9gMuLVpmePbxNYujvPy8CuBu/PwOk3Ln5C3tfJ2m131/9Xa/BiWtYBszUXEEkk/Ao4Gnuhv+uz6iHgIQNJfgUat8zlAcxv4RXkdd0u6HdgOeAPwUklvz9NsBLwEeBK4LiLua7O+3YArI+LRvM7zgL2AC/Pr7fZ630A6ApmVXx8PvCi/dhJwfX6/RzbNc12komiQfnT3JNVIeSlwTd67Xhf4U+t77KBRQ2kO6QfvKeApSUslPbdLjH8F5kXEbXn+G0lJsKH5/b5F0idJdbo2of8CbreQjiAuAf6nn2lb7Une5hHxgKR2zV170X57PQwsk/TfwKWkZFfExXl9tysXLiO911PzUcLTpKTzzwN8L7YGnADWTt8idVHZXEBvBbnJL/8jr9f02lMt0z3VOk/WXDhKTc+PiIiZzQEo9TmwtEuMA630KtKeeruigJuSjnxGk45KGomvtdBV4/mMiDisw3q6xdy8XVq32ahOMebmkObpl9Om+VWpuuipwKSIuF/SVPr/H30z8FrS0cfxkrZtM02nz75IITDRYXtJ2pV0hHMw8DFW3VnopHk7NNb/flIXoNvl5c7Dv0094XMAaxfByh6MLgQ+2PTaAlJ7M8CBpD25gToYVvaA9jJSWekrgI8on+iVtJX67zD7WlK9/+fk+d5B6sSim8tJna+Myet5XtMe5BnAfwA/Ab7WNM+rJU3MP3pvJ/WHfHVe9wvzctZX6vN1MBrJrF2Mm7ZM0+oJ0pECpB+95cAjeRse0mEe8vIFPD8irgI+S2r2e27LMqHzZ381absgaTNgnzar+T1ttldOVhtGqn7778AO/by3tm8h/x1L7qZU0mtJTT+N+cd1md8GyVl27dK8R3cyaa+sMe4MYIakN5B+qDrt6XbbK1wg6VpS++8REfGUpNNJ5wRuk7SM1I7ctR/SiFigdHnqH/P6Lo+IRtNL2/VHxHRJrwRm5fU8CbxL0gHAsni24/NrJE3Oy7mBdF7glcDVkfvHlfRh4GJJQdoJ+g9SffdORwz9bZdoinGb1hhJe+Cd5v8xcLakx4DdgR8CdwB3A9f1s/51gAuUyimvA5weEQ9Lmg78PJ/4/TidP/sLgH0l/Rm4i9TW3/qeFnbYXo/ncaPzuGPaxNcaR6ft+xPgckk3kz6z2/O6H1A6wX8b6fzTp9uswwbB5aBtrZSboD4ZEQdVHYvZcOUmIDOzmvIRgJlZTfkIwMysppwArFTNd5QWnH7vfBKz8Xy0pBvLiW7lOi6RtNEaznuYpNP6mWZNa/PMU0llPVq3s9WTrwKy4ai5XXJP0uWK5a0s4sDBLqKf1ycDS0iXvw7lcgfL7b815yOAmpG0oaRfKdXumS3p0Dy+bQ2avAd/ilINoLmSdpF0kVI9oK/mabbM801Tqg0zvd29AOpcU+b/KdUh+iPQWi1yf+DSpnWcqVS75ieS9lOqK3NXYw87x3dtfn835EtHG3vqP8t7+3+TdGpTXPMkbdLtfTTvjUvaSdLMljgb7+9P+f1dJemf1L42z+Y5jpuV6ujsneffNL+fmyR9nw73DnTZjp0+w1dI+n2+pHKWpMYd1Bsq1Q+6U9KFHda1R17enyR9TdKcPL5bDaiZ7ZYrabf82dwi6beStmi3TuuhqmtR+NHbB+nGn9Obnjfqx3SqQTMT+FIePgq4j1XrBm1GunFnBbBLnu77wGea5u9WU2YM8L/Alnn8ucDFTbH8KU/TqKfz8jz+BuAHefggnq1TNL5p3n2B6Xn4MFJJhrGkGj73AFvl1+4ilV3o9j7uAjbJwzsBv21a7rfbbMMPNo1vrYnzM+A1efgFwF/z8BnAZ/PwG0g3hW3S8vm1245f7OczvBk4IA+vk7fn3qR7Np6Xx/8B2LvN9+UOYKc8/EVybR6614Babbmkm89uACbk8YcC51T9/1D3h5uA6udm4CuSvgJcGhG/y+O71aC5JP+dA8yJVesGPZ/0D39vRFyfpzuPVIiuWbuaMtcB2wN3RMQ9TfN+KC9/C+ChiHgy78zOi4g78nS3AY3aNXOAiXl4M0kXkG5OW0H6sWv4TUQ8kZd9W57nblbd0+70PoqUrtha0imk0hSjgfkdpns98KLGHjqwntI5iL2A/wSIiCskPdJm3k7bEdp8hkp3Ik+IiBl5ucuB5XnV10XEA3nem1m1PlHj7uD1IqJxDuYCUrKF/mtAtS53Uo771znuUaSCglYhJ4CaiYi/SNoZOAD4glIBsFPoXoOmvxo4bVfV8rxtTZkcS6cf1/1Jd662xtEaS3McX8rr+W5ufpnZYf629XjaaLyP5nWM6TDtd0h745fmZp2pXZa5S/4xXknSipbp2m2XTtuxWx2hTtt3TbZHcxydakC1W66AmyNi7wGsw0rmcwA1o9T71NKIOBf4BrALA6xB08EL8485pNo+rSdu29aUIdUTemljfJ63YX9SpcmV4ReIYyypSQlSkbGB6vQ+5pOafgDe1mHeMcD9efh9TeNba+JcCazs6UrPFnC7mlSyGkn7ker6tOq0Hdt+hhHxIPCApAPz9KPVf60m8ryLSBU/G1dxvb3p5csZWA2o2aSKSarnAAABSklEQVRtu0Oefh1JLy8Sh5XHCaB+JgE3KpUs/hypeuWjPFuD5lL6r0HT7rU7gCMl3UpqWvl28zQRsRBo1JS5Ka9jm4h4knSS9DdKJ4EXwcpexLaOiDs7rK9TXN8ATpZ0Hd0L3nVaVqf3cSJwhlItpNY99YYvAr/I7+OhpvHTgXfnk7t7kN7vfpLm5JOqR+fpPgcckLfPwcC9qwXdeTt2+wzfDXxW0mxS/aXN+9kezf4V+ImkP5GaeRqVVk8nnQO6TdIted3tWhQan/8yUgI5I8d9M+ncgFXIdwLboOWmlksiYvshXOYewHsiomddApbxPkY6SWObzpt8mnSy3t00riV8DsCGypDuSUTENaTyzb3mPaJVHSTps6SmtfmkowlbS/gIwMyspnwOwMysppwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMaur/AMUSNv+qtJ0lAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fbf090d7780>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "temp = transformed_subset.exp_source.apply(lambda x: len(x))\n",
    "plt.hist(temp, bins = 10, histtype = 'stepfilled')\n",
    "ax.set_xlabel('Number of experimental studies that\\nsampled/manipulated each gene')\n",
    "ax.set_ylabel('Number of unique genes')\n",
    "plt.savefig('simulations_output/num_exp_source_per_unique_gene.png', bbox_inches = \"tight\", dpi = 300) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
