{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregate selective studies data\n",
    "\n",
    "Define **selective studies** as studies that \n",
    "- Have at these one of the following tags:\n",
    "    > ['knock_out','gene_silence','RT_qPCR','gene_knockout','genetic_transformation']\n",
    "- have at least 1 gene manipulated.\n",
    "- are not one of the studies in `match_summary.csv`\n",
    "\n",
    "##### Output file path: `lipid_selection/data/intermediate_data_02/selective_studies.pk`\n",
    "Output format:\n",
    "\n",
    "| transcript_id | s_num_detected | s_num_sampled |s_num_no_effect | s_proportion | s_source                 | annotation_version |\n",
    "|---------------|------------------|-----------------|------------------|----------------|----------------------------|--------------------|\n",
    "| `<string>`    | `<int>`          | `<int>`         | `<float>`        | `<float>`      | `'Firstauthor_year'` | `v5.5`             |\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##### Output file path: `lipid_selection/data/intermediate_data_02/selective_studies_reduced.pk`\n",
    "Output format:\n",
    "\n",
    "| transcript_id | s_num_detected | s_num_sampled |s_num_no_effect | s_proportion | s_source                 | annotation_version |\n",
    "|---------------|------------------|-----------------|------------------|----------------|----------------------------|--------------------|\n",
    "| `<string>`    | `<int>`          | `<int>`         | `<float>`        | `<float>`      | `['Firstauthor_year',...]` | `v5.5`             |\n",
    "\n",
    "**`s_proportion`:** `<float>` \n",
    "> `s_proportion` = `s_num_detected`/ `num_selective_source`\n",
    "\n",
    "**`num_selective_source`:** `<int>` number of selective studies\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import pickle\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"../../data/raw_data/Summary_of_primary_data.xlsx\", 'Sheet1')\n",
    "df = df[['Genus', 'Species','First_author','Publication_year','Type_of_study','Genes_sampled','Genes_manipulated', 'Genes_with_effect',\"Genes_no_effect\"]]\n",
    "match_summary = pd.read_csv(\"../../data/intermediate_data_02/match_summary.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Genus      Species First_author  Publication_year  \\\n",
      "0  Nannochloropsis          NaN        Zheng              2013   \n",
      "1    Chlamydomonas  reinhardtii           Lv              2013   \n",
      "2    Chlamydomonas  reinhardtii           Lv              2013   \n",
      "3      Fistulifera      solaris       Tanaka              2015   \n",
      "4      Fistulifera      solaris        Osada              2017   \n",
      "\n",
      "     Type_of_study Genes_sampled                     Genes_manipulated  \\\n",
      "0  transcriptomics           NaN                                   NaN   \n",
      "1  transcriptomics           NaN                                   NaN   \n",
      "2        knock_out           NaN  Cre05.g248150.t1; Cre06.g299050.t1.2   \n",
      "3  transcriptomics           NaN                                   NaN   \n",
      "4  transcriptomics           NaN                                   NaN   \n",
      "\n",
      "    Genes_with_effect   Genes_no_effect  \n",
      "0                 NaN               NaN  \n",
      "1                 NaN               NaN  \n",
      "2  Cre06.g299050.t1.2  Cre05.g248150.t1  \n",
      "3                 NaN               NaN  \n",
      "4                 NaN               NaN  \n",
      "  first_author    year  matches  mismatches    total  proportion_of_matches\n",
      "0        Hemme  2014.0   1894.0        86.0   1980.0               0.956566\n",
      "1     Bajhaiya  2016.0  17736.0         1.0  17737.0               0.999944\n",
      "2   Goodenough  2014.0     66.0         2.0     68.0               0.970588\n",
      "3        Blaby  2013.0    591.0        51.0    642.0               0.920561\n",
      "4        Boyle  2012.0     24.0         1.0     25.0               0.960000\n"
     ]
    }
   ],
   "source": [
    "print(df.head())\n",
    "print(match_summary.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the types of studies?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "types_of_studies = []\n",
    "\n",
    "for study in list(df.Type_of_study.unique()):\n",
    "    try: \n",
    "        study = study.split(\"; \")\n",
    "        for i in study: \n",
    "            if not i in types_of_studies: types_of_studies.append(i)\n",
    "    except: pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['transcriptomics', 'knock_out', 'gene_silence', 'proteomics', 'RT_qPCR', 'gene_knockout', 'genomics', 'genetic_transformation']\n"
     ]
    }
   ],
   "source": [
    "print(types_of_studies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset data to selective studies only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "re.compile('knock_out|gene_silence|RT_qPCR|gene_knockout|genetic_transformation')\n",
      "re.compile('Hemme|Bajhaiya|Goodenough|Blaby|Boyle|Juergens|Kwak|Gargouri')\n"
     ]
    }
   ],
   "source": [
    "pattern_study = re.compile(\"|\".join(['knock_out','gene_silence','RT_qPCR','gene_knockout','genetic_transformation']))\n",
    "print(pattern_study)\n",
    "\n",
    "pattern_author = re.compile(\"|\".join(list(match_summary.first_author)))\n",
    "print(pattern_author)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Genus      Species First_author  Publication_year Type_of_study  \\\n",
      "2   Chlamydomonas  reinhardtii           Lv              2013     knock_out   \n",
      "9   Chlamydomonas  reinhardtii        Siaut              2011     knock_out   \n",
      "10  Chlamydomonas  reinhardtii         Wang              2009     knock_out   \n",
      "11  Chlamydomonas  reinhardtii   Moellering              2009  gene_silence   \n",
      "16  Chlamydomonas  reinhardtii           Li              2010     knock_out   \n",
      "\n",
      "   Genes_sampled                     Genes_manipulated  \\\n",
      "2            NaN  Cre05.g248150.t1; Cre06.g299050.t1.2   \n",
      "9            NaN                STA1; STA6; STA7; STA8   \n",
      "10           NaN                                  STA6   \n",
      "11           NaN                                  MLDP   \n",
      "16           NaN         STA6; STA7; STA1; STA11; ARG7   \n",
      "\n",
      "                Genes_with_effect   Genes_no_effect  \n",
      "2              Cre06.g299050.t1.2  Cre05.g248150.t1  \n",
      "9          STA1; STA6; STA7; STA8               NaN  \n",
      "10                           STA6               NaN  \n",
      "11                           MLDP               NaN  \n",
      "16  STA6; STA7; STA1; STA11; ARG7               NaN   \n",
      "\n",
      " (23, 9)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "subset = df[df.Type_of_study.str.contains(pattern_study)==True][df.Genes_manipulated.isnull()==False][df.First_author.str.contains(pattern_author)==False]\n",
    "print(subset.head(), '\\n'*2,subset.shape)\n",
    "num_selective_source = subset.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Genus      Species First_author  Publication_year  \\\n",
      "2   Chlamydomonas  reinhardtii           Lv              2013   \n",
      "9   Chlamydomonas  reinhardtii        Siaut              2011   \n",
      "10  Chlamydomonas  reinhardtii         Wang              2009   \n",
      "11  Chlamydomonas  reinhardtii   Moellering              2009   \n",
      "16  Chlamydomonas  reinhardtii           Li              2010   \n",
      "\n",
      "      Type_of_study Genes_sampled                       Genes_manipulated  \\\n",
      "2      (knock_out,)           NaN  (Cre05.g248150.t1, Cre06.g299050.t1.2)   \n",
      "9      (knock_out,)           NaN                (STA1, STA6, STA7, STA8)   \n",
      "10     (knock_out,)           NaN                                 (STA6,)   \n",
      "11  (gene_silence,)           NaN                                 (MLDP,)   \n",
      "16     (knock_out,)           NaN         (STA6, STA7, STA1, STA11, ARG7)   \n",
      "\n",
      "                  Genes_with_effect      Genes_no_effect  \n",
      "2             (Cre06.g299050.t1.2,)  (Cre05.g248150.t1,)  \n",
      "9          (STA1, STA6, STA7, STA8)                  NaN  \n",
      "10                          (STA6,)                  NaN  \n",
      "11                          (MLDP,)                  NaN  \n",
      "16  (STA6, STA7, STA1, STA11, ARG7)                  NaN  \n"
     ]
    }
   ],
   "source": [
    "for col in df.columns[4:]:\n",
    "    subset[col] = subset[col].apply(lambda x: tuple(x.split(\"; \")) if isinstance(x,str) else np.NaN) \n",
    "print(subset.head())  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Genus', 'Species', 'First_author', 'Publication_year', 'Type_of_study',\n",
      "       'Genes_sampled', 'Genes_manipulated', 'Genes_with_effect',\n",
      "       'Genes_no_effect'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(subset.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reaarrange subset to target datframe format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert 20201019_sampled_genes_key.csv to dictionary format with `symbol` as keys and `transcript_id` as values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                              source symbol  \\\n",
      "0  Sim, L., Beeren, S. R., Findinier, J., Dauvill...   STA8   \n",
      "1                                        Phytozome12   APL1   \n",
      "2                                        Phytozome12  PEPC1   \n",
      "3                                              Query   NRR1   \n",
      "4                                              Query   SQD1   \n",
      "\n",
      "        transcript_id  \n",
      "0  Cre17.g698850.t1.1  \n",
      "1  Cre16.g683450.t1.1  \n",
      "2       Cre16.g673852  \n",
      "3  Cre16.g673250.t1.1  \n",
      "4  Cre16.g656400.t1.2  \n",
      "[('STA7', 'Cre03.g155001.t1.1'), ('PDC2', 'Cre02.g099850.t1.1'), ('DGAT2', 'Cre03.g205050.t1.2'), ('FEA1', 'Cre12.g546550.t1.1'), ('CDD12', 'Cre13.g590500')]\n"
     ]
    }
   ],
   "source": [
    "sampled_genes_key = pd.read_csv(\"../../analysis/01_generate_intermediate_data_01/20201019_sampled_genes_key.csv\")\n",
    "print(sampled_genes_key.head())\n",
    "#print()\n",
    "key = dict(zip(sampled_genes_key.symbol, sampled_genes_key.transcript_id))\n",
    "print(list(key.items())[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset = pd.DataFrame(columns = ['transcript_id','s_num_sampled', 's_num_manipulated', 's_num_detected', 's_num_no_effect', 's_source'])\n",
    "for index, row in subset.iterrows():\n",
    "    source = row.First_author + \"_\" + str(row.Publication_year)\n",
    "    for i in range(0,4): \n",
    "        if type(row[i+5]) != float:\n",
    "            for gene in row[i+5]: #Genes_sampled\n",
    "                output = [0]*(i)+[1]+[0]*(3-i)\n",
    "                try: \n",
    "                    gene = key[re.sub(r\" |;\",\"\", gene)]\n",
    "                except: gene = re.sub(r\" |;\",\"\", gene)\n",
    "                output = pd.Series([gene]+output+[source], index = transformed_subset.columns)\n",
    "                transformed_subset = transformed_subset.append(output, ignore_index = True)\n",
    "                #print(output)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset['annotation_version'] = 'v5.5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>s_num_sampled</th>\n",
       "      <th>s_num_manipulated</th>\n",
       "      <th>s_num_detected</th>\n",
       "      <th>s_num_no_effect</th>\n",
       "      <th>s_source</th>\n",
       "      <th>annotation_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cre05.g248150.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre06.g299050.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre06.g299050.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cre05.g248150.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cre13.g567950.t1.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Siaut_2011</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  s_num_sampled  s_num_manipulated  s_num_detected  \\\n",
       "0  Cre05.g248150.t1.2            0.0                1.0             0.0   \n",
       "1  Cre06.g299050.t1.2            0.0                1.0             0.0   \n",
       "2  Cre06.g299050.t1.2            0.0                0.0             1.0   \n",
       "3  Cre05.g248150.t1.2            0.0                0.0             0.0   \n",
       "4  Cre13.g567950.t1.2            0.0                1.0             0.0   \n",
       "\n",
       "   s_num_no_effect    s_source annotation_version  \n",
       "0              0.0     Lv_2013               v5.5  \n",
       "1              0.0     Lv_2013               v5.5  \n",
       "2              0.0     Lv_2013               v5.5  \n",
       "3              1.0     Lv_2013               v5.5  \n",
       "4              0.0  Siaut_2011               v5.5  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/intermediate_data_02/selective_studies.pk', 'wb') as f:\n",
    "    pickle.dump(transformed_subset, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = transformed_subset.groupby(['transcript_id'])[['s_num_sampled', 's_num_manipulated', 's_num_detected', 's_num_no_effect']].sum()\n",
    "source = transformed_subset.groupby(['transcript_id'])[['s_source']].agg(lambda x: list(np.unique(list(itertools.chain(x)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_num_sampled</th>\n",
       "      <th>s_num_manipulated</th>\n",
       "      <th>s_num_detected</th>\n",
       "      <th>s_num_no_effect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    s_num_sampled  s_num_manipulated  s_num_detected  \\\n",
       "transcript_id                                                          \n",
       "Cre01.g021251.t1.1            0.0                1.0             1.0   \n",
       "Cre01.g038550.t1.1            1.0                1.0             1.0   \n",
       "Cre01.g045903.t1.1            3.0                0.0             1.0   \n",
       "Cre01.g053000.t1.2            1.0                0.0             1.0   \n",
       "Cre01.g053150.t1.2            1.0                0.0             1.0   \n",
       "\n",
       "                    s_num_no_effect  \n",
       "transcript_id                        \n",
       "Cre01.g021251.t1.1              0.0  \n",
       "Cre01.g038550.t1.1              0.0  \n",
       "Cre01.g045903.t1.1              1.0  \n",
       "Cre01.g053000.t1.2              0.0  \n",
       "Cre01.g053150.t1.2              0.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>[Li_2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>[Iwai_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               s_source\n",
       "transcript_id                                          \n",
       "Cre01.g021251.t1.1                            [Li_2010]\n",
       "Cre01.g038550.t1.1                          [Iwai_2014]\n",
       "Cre01.g045903.t1.1  [Iwai_2014, Msanne_2012, Sato_2014]\n",
       "Cre01.g053000.t1.2                          [Sato_2014]\n",
       "Cre01.g053150.t1.2                          [Sato_2014]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_num_sampled</th>\n",
       "      <th>s_num_manipulated</th>\n",
       "      <th>s_num_detected</th>\n",
       "      <th>s_num_no_effect</th>\n",
       "      <th>s_source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    s_num_sampled  s_num_manipulated  s_num_detected  \\\n",
       "transcript_id                                                          \n",
       "Cre01.g021251.t1.1            0.0                1.0             1.0   \n",
       "Cre01.g038550.t1.1            1.0                1.0             1.0   \n",
       "Cre01.g045903.t1.1            3.0                0.0             1.0   \n",
       "Cre01.g053000.t1.2            1.0                0.0             1.0   \n",
       "Cre01.g053150.t1.2            1.0                0.0             1.0   \n",
       "\n",
       "                    s_num_no_effect                             s_source  \n",
       "transcript_id                                                             \n",
       "Cre01.g021251.t1.1              0.0                            [Li_2010]  \n",
       "Cre01.g038550.t1.1              0.0                          [Iwai_2014]  \n",
       "Cre01.g045903.t1.1              1.0  [Iwai_2014, Msanne_2012, Sato_2014]  \n",
       "Cre01.g053000.t1.2              0.0                          [Sato_2014]  \n",
       "Cre01.g053150.t1.2              0.0                          [Sato_2014]  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset = pd.merge(sample, source, left_index = True, right_index= True)\n",
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset['s_proportion'] = transformed_subset.s_num_detected / num_selective_source\n",
    "transformed_subset['annotation_version'] = 'v5.5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_num_sampled</th>\n",
       "      <th>s_num_manipulated</th>\n",
       "      <th>s_num_detected</th>\n",
       "      <th>s_num_no_effect</th>\n",
       "      <th>s_source</th>\n",
       "      <th>s_proportion</th>\n",
       "      <th>annotation_version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cre01.g021251.t1.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g038550.t1.1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g045903.t1.1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053000.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cre01.g053150.t1.2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    s_num_sampled  s_num_manipulated  s_num_detected  \\\n",
       "transcript_id                                                          \n",
       "Cre01.g021251.t1.1            0.0                1.0             1.0   \n",
       "Cre01.g038550.t1.1            1.0                1.0             1.0   \n",
       "Cre01.g045903.t1.1            3.0                0.0             1.0   \n",
       "Cre01.g053000.t1.2            1.0                0.0             1.0   \n",
       "Cre01.g053150.t1.2            1.0                0.0             1.0   \n",
       "\n",
       "                    s_num_no_effect                             s_source  \\\n",
       "transcript_id                                                              \n",
       "Cre01.g021251.t1.1              0.0                            [Li_2010]   \n",
       "Cre01.g038550.t1.1              0.0                          [Iwai_2014]   \n",
       "Cre01.g045903.t1.1              1.0  [Iwai_2014, Msanne_2012, Sato_2014]   \n",
       "Cre01.g053000.t1.2              0.0                          [Sato_2014]   \n",
       "Cre01.g053150.t1.2              0.0                          [Sato_2014]   \n",
       "\n",
       "                    s_proportion annotation_version  \n",
       "transcript_id                                        \n",
       "Cre01.g021251.t1.1      0.043478               v5.5  \n",
       "Cre01.g038550.t1.1      0.043478               v5.5  \n",
       "Cre01.g045903.t1.1      0.043478               v5.5  \n",
       "Cre01.g053000.t1.2      0.043478               v5.5  \n",
       "Cre01.g053150.t1.2      0.043478               v5.5  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset['transcript_id']= transformed_subset.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_subset.index = range(len(transformed_subset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_num_sampled</th>\n",
       "      <th>s_num_manipulated</th>\n",
       "      <th>s_num_detected</th>\n",
       "      <th>s_num_no_effect</th>\n",
       "      <th>s_source</th>\n",
       "      <th>s_proportion</th>\n",
       "      <th>annotation_version</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Li_2010]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g021251.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Iwai_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g038550.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[Iwai_2014, Msanne_2012, Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g045903.t1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g053000.t1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[Sato_2014]</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g053150.t1.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   s_num_sampled  s_num_manipulated  s_num_detected  s_num_no_effect  \\\n",
       "0            0.0                1.0             1.0              0.0   \n",
       "1            1.0                1.0             1.0              0.0   \n",
       "2            3.0                0.0             1.0              1.0   \n",
       "3            1.0                0.0             1.0              0.0   \n",
       "4            1.0                0.0             1.0              0.0   \n",
       "\n",
       "                              s_source  s_proportion annotation_version  \\\n",
       "0                            [Li_2010]      0.043478               v5.5   \n",
       "1                          [Iwai_2014]      0.043478               v5.5   \n",
       "2  [Iwai_2014, Msanne_2012, Sato_2014]      0.043478               v5.5   \n",
       "3                          [Sato_2014]      0.043478               v5.5   \n",
       "4                          [Sato_2014]      0.043478               v5.5   \n",
       "\n",
       "        transcript_id  \n",
       "0  Cre01.g021251.t1.1  \n",
       "1  Cre01.g038550.t1.1  \n",
       "2  Cre01.g045903.t1.1  \n",
       "3  Cre01.g053000.t1.2  \n",
       "4  Cre01.g053150.t1.2  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transformed_subset.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export `transformed_subset` as `lipid_selection/data/intermediate_data_02/selective_studies.pickle`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/intermediate_data_02/selective_studies_reduced.pk', 'wb') as f:\n",
    "    pickle.dump(transformed_subset, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is the distribution of number of s_source and s_num_sampled/manipulated for the genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "fig, ax = plt.subplots()\n",
    "plt.hist(transformed_subset.s_proportion, bins = 10, histtype = 'stepfilled')\n",
    "ax.set_xlabel('Proportion of datasets that detected effect of a unique gene')\n",
    "ax.set_ylabel('Number of unique genes')\n",
    "plt.savefig('simulations_output/distribution_of_s_proportion.png', bbox_inches = \"tight\", dpi = 300)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEbCAYAAADTZlM/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmcHWWd7/HPN2whgWgIDA6LoCA4LFGEKKtsIojI1YCIKzLMyEUFBC+4vC5GQcdxARFGZXQAjciuImGNSBoMIlsISQiCyJbINQSIkIQlkPzuH8/TycnJOd3V6a5T3db3/XqdV9epruV3qpPzq3qeqt+jiMDMzOpnWNUBmJlZNZwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMaqrUBCBpHUl3SZom6UFJZ+X5W0r6g6QZki6RtGaZcZiZ2apKTQAR8TLwzoh4G7AdsLukfYFzgG9FxFhgHvDZMuMwM7NVld4EFBEv5sl18v7mAbtGxG/y/IuAQ8qOw8zMVlZ6ApA0TNK9wN+ALmAB8HTDInOBTcuOw8zMVlZ623tELAN2kjQKuBGYXvY+zcysdx3rfI2I5yVdB7wR2LDhV5uRrgJWIcmFiszMVkNEqLdlyr4LaIyk9fL0usABwL3AHyW9Py/2MeD69luJCl/PMGLEaCKix9eECRN6XWYwvBxnvWJ0nPWNs6iyrwA2ASZKAhgOXBwR10qaDVws6XRgNnBKyXGYmVmTUhNARMwEdmox/1FgtzL3bWZmPfOTwANgn332qTqEQhznwBkKMYLjHGhDJc6i1Jf2ok5LncBVxvcsI0ZszeLFz1YYg5lZ30giqu4ENjOzwcsJwMysppwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMasoJwMysppwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMasoJwMysppwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMasoJwMysppwAzMxqygnAzKymnADMzGrKCcDMrKacAMzMasoJwMyspkpNAJI2k3SLpJmS/iTplDx/gqS5kqbl10FlxmFmZqtas+TtvwJ8JiJmSVoPmCbpxvy7syLirJL3b2ZmbZSaACJiHjAvTy+SNAPYNP9aZe7bzMx61rE+AElbArsAU/OsT0uaLennkkZ3Kg4zM0s6kgBy888VwIkRsRD4L2DriNgOeAQ4txNxmJnZCmX3ASBpTeBK4BcR8RuAiHimYZHzgCntt/DVhul98svMzLp1dXXR1dXV5/UUEQMfTeMOpInA0xFxcsO8jSJifp4+Htg3Isa3WDeg3Ph69iwjRmzN4sXPVhiDmVnfSCIieu1nLfUKQNIewEeBmZLuJX2bfxn4qKSxwFrAE8AxZcZhZmarKv0KoD98BWBm1ndFrwD8JLCZWU05AZiZ1ZQTgJlZTTkBmJnVlBOAmVlNOQGYmdWUE4CZWU05AZiZ1ZQTgJlZTTkBmJnVVJ8SgKTRkt5WVjBmZtY5vSYASbdKGilpQ2A68CNJ3y8/NDMzK1ORK4D1ImIxMB64ICLeAexbblhmZla2IglgLUkbAYcB1+V5g7eEqJmZFVIkAfwH0AU8EhF3SdqCNIyjmZkNYR4PoEceD8DMhp4BGw9A0naSpkp6oOH9hIEI0szMqlOkCegC4GTgJYCImA18sMygzMysfEUSwPCIuLNp3tIygjEzs84pkgAWSNqK3Bgv6RDAjeJmZkPcmgWWORb4GfAvkh4H5gNHlhqVmZmVrtcEEBEPAXtKGkO6a+jp8sMyM7Oy9ZoAJI0gdfpuDgyT0p1FEXF6uaGZmVmZijQBXQs8BdyDO3/NzP5hFEkAG0WEa/+Ymf2DKXIX0FRJ25ceiZmZdVSRK4B3Av8q6VHgZUBARMTYUiMzM7NSFUkA7yk9CjMz67hem4Ai4nFga+CAPL24yHpmZja4FSkG9x/A54BTG9a5pMjGJW0m6RZJMyX9SdKpef5oSZMl3SfpBkmvWd0PYGZmq6fImfz7gUNJZ/5ExFPA8ILbfwX4TETsCOwCHCNpLPA14LqIeAtwA+BnCszMOqxIAng10qAB3bWAhgNrF9l4RMyLiFl5ehEwE9gMeC/w87zYRfm9mZl1UJEEcIWk/wZeK+loYDLw077uSNKWpKuA35OeLXgGIJeW2Kiv2zMzs/4pUgvoDEnvA5YAOwPfiYhJfdmJpPWAK4ATI2JhGumrqK82TO+TX2Zm1q2rq4uurq4+r1f6kJCS1gSuAW6IiLPzvIeBd0TEM5I2BG6PiDe1WNdDQpqZ9dFADgm5UNLzTa95kq6TtE2BWC4AZnd/+WfXAR/P0x8Hri+wHTMzG0C9XgFIOgN4HLiM9BTwB4FtgGnASRGxaw/r7gHcSur87e5I/jJwZ97exsDfgCMi4u8t1vcVgJlZHxW9AiiSAO6OiF2a5t0VEeMkzY6I7foZa0/7dgIwM+ujAWsCAtaQdJikYfl1GCs6j5f0K0ozM6tMkQRwJGlYyOeAv+fpj+TnAY4vMTYzMytR6XcB9YebgMzM+m4gm4DMzOwfkBOAmVlNOQGYmdVUkQfBNpF0kaTJ+f22kj5VfmhmZlamIlcAFwFXkx7aAngY3/1jZjbkFUkAYyLicmAZQEQsBV4tNSozMytdkQTwgqQxrBgPYCf8AJiZ2ZBXZFD4k4Ebga0k3QK8nlQPyMzMhrAi4wHcLmk3YCypGNyMiPAVgJnZENdrApD0iaZZ2+WnzCaWFJOZmXVAkSagcQ3Tw4H9SaWgnQDMzIawIk1AK93yKWl9Ui1/MzMbwlbnSeCXgG0HOhAzM+usIn0Ak1hRknMYsD3wmzKDMjOz8hXpA/huw/QyYG5EPFpSPGZm1iFF+gBu6UQgZmbWWUWagBbSelQWARERowY8KjMzK12RJqCzgTnAJaQv/Q8Bm0XEhDIDMzOzcvU6JKSkuyNil97mlcFDQpqZ9d1ADgk5TNKRktaQNEzSkQXXMzOzQazIF/mHgE8CC/LrqDzPzMyGsCJ3Af0ZOKgDsZiZWQe1TQCSTo2Ib0s6lxYN8RFxQqmRmZlZqXq6Angg/7y7E4GYmVln9XoXUJV8F5CZWd8N2F1AkraXdKGkmyTd3P0qGMT5kuZJmtEwb4KkuZKm5Zf7F8zMKlDkQbBfAt8HfgQs7eP2LwTOZdWxA86KiLP6uC0zMxtARRLA8xHxo9XZeERMlbRFi1/1emliZmblKvIcwHWSjpX0z5I26H71c7+fljRb0s8lje7ntszMbDUUKQXRqvRzRMQbC+0gXQFMioix+f0Y4NmICElfA7aKiI+1WdedwGZmfVS0E7jIg2BvGJiQlm/vmYa35wFTel7jqw3T++SXmZl16+rqoqurq8/rFbkC+ESr+RFRaFB4SVuSrgB2zO83ioj5efp4YN+IGN9mXV8BmJn10YBdAQDjGqaHA/sD01j1zp5WQVxMOmUfI+kJYAKwn6SxwFrAE8AxBWIwM7MB1ucHwSStD1wWEQeXE9JK+/IVgJlZHw1kOehmLwHbrsZ6ZmY2iBQZEnISK07DhwHbA78pMygzMytfkT6A7zZMLwPmRkSrW0PNzGwIKXIb6C2dCMTMzDrLQzuamdWUE4CZWU21TQCSfpd/fqtz4ZiZWaf01AewuaTdgUMlXUpTBc+ImFZqZGZmVqqeEsD/BU4DNgOaa/cHsF9ZQZmZWfmK1AI6LSLO6FA8zfv2k8BmZn00kNVAz5B0BLBXnnVLRFzZ3wDNzKxaRZ4E/h6wA3BJnvXvkvaIiJNKjczMzEpV5EngdwM7RG4rkvRTYFaZQZmZWfmKPAcgYFTD+/XxmL5mZkNekSuA7wD3S7qJ9MW/L/CVUqMyM7PSFRoPQNLmwK757e0RMbfUqFbs13cBmZn10UCOCEZEzAHm9DsqMzMbNFwLyMysppwAzMxqqscEIGmYpPs7FYyZmXVOjwkgIpYBD0napEPxmJlZhxTpBB4DPCjpTmBx98yIOLS0qMzMrHRFEsBppUdhZmYdV2hMYElbA1tFxI2S1i2ynpmZDW693gUk6XjgMuCHedbGwKQygzIzs/IVuQ30OGA34HmAiHgMGF1iTGZm1gFFEsCSiFjS/UbSMGDt8kIyM7NOKJIAfi/py8C6kvYFLgauKzcsMzMrW5EE8DlgIfBn4GTgNuCUIhuXdL6keZJmNMwbLWmypPsk3SDpNasTuJmZ9U/RaqDDSaOCBTArIl4utHFpT2ARMDEixuZ55wCPRMTZkj4HvCEiTmyzvquBmpn1UdFqoEXuAvoA8DjwXeAs4DFJ7y8SRERMBRY0zX4v8PM8fVF+b2ZmHVZ0QJh35Lt/kLQlcBNw1Wruc8OIeAYgIp6WtNFqbsfMzPqhSB/A091f/rD8NtD5ZQVkZmad0fYKQNL4PHmvpGuAK0kN8ocB9/Zjn/MljYmIZyRtCDzV8+JfbZjeJ7/MzKxbV1cXXV1dfV6vbSewpAt7WjEiji60g9RkNCkidszvGzuBTyJ1Ap/QZl13ApuZ9VHRTuBCdwH1I4iLSafsY4B5wARS38HlpJISfwOOiIi/t1nfCcDMrI8GLAHkQnAnApvT0GfQiXLQTgBmZn03kIPCX0MqBPcrYFl/AzMzs8GhSAJYEBHnlB6JmZl1VJEE8ANJp5Hu/V/+BHBETCstKjMzK12RBLAD8HHgXaxoAgpgv7KCMjOz8hVJAIeRbtVc0uuSZmY2ZBR5EngGsH7ZgZiZWWcVuQIYA/xZ0l2s3AdQ+m2gZmZWniIJYELpUZiZWcf1mgAi4pZOBGJmZp3VawKQtJAVj+OuDawFLI6IUWUGZmZm5SpyBbC8A1iSgEOA3csMyszMylfkLqDlIpkEHFhSPGZm1iFFmoDGN7wdBuwC9FpkyMzMBrcidwG9r2F6GTAXOLiccMzMrFOK9AEUGvjFzMyGlp6GhPxKD+tFRJxRQjxmZtYhPV0BLG4xbyRwDOnpYCcAM7MhrG0CiIgzu6clrU8aFexo4FLgzHbrmZnZ0NBjH4CkDYCTgY8CPwPeFhELOhGYmZmVq6c+gO8A44EfAztGxKKORWVmZqVrOyi8pGWk6p+vsvLI7CJ1ApdeCsKDwpuZ9V2/B4WPiD49JWxmZkOLv+TNzGrKCcDMrKacAMzMasoJwMysppwAzMxqygnAzKymipSDLoWkx4DnSCWmX4mIt1cVi5lZHVWWAEhf/Pu4tISZWTWqbAJSxfs3M6u1Kr+AlwGTJd0n6bMVxmFmVktVNgHtFhFPSdoIuEHSAxHxuwrjMTOrlcoSQEQ8lX/Ol3QlMA5okQC+2jC9T36ZmVm3rq4uurq6+rxe22qgZZI0glRR9EVJI4HrgDMj4uqm5VwN1Mysj/pdDbRkGwNX5ZLTI4BLm7/8zcysXJUkgIh4FHhLFfs2M7PEt2GamdWUE4CZWU05AZiZ1ZQTgJlZTTkBmJnVlBOAmVlNOQGYmdWUE4CZWU05AZiZ1ZQTgJlZTVVZDtoKmjLlVn74wwurDoNx43bi1FNPqDoMMxsglVQDLcrVQJOjjjqOiRMXAftVGMWzvPa132PBgrkVxmBmRQz2aqDWZ7sDR1e4/7nA9yrcv5kNNPcBmJnVlBOAmVlNOQGYmdWUE4CZWU25E9gKW7p0CTNmzKg0hmHDhrH99tsj9XqDQ6keeeQRFi1aVGkMa6yxBtttt13lx8KGLicAK2g0Eduz114fqzSKl16awyWXnM/48eMri2HOnDm8+c07sO66W1cWA8BLLz3O1VdfzoEHHlhpHDZ0OQFYQSNZtGhK1UEwcuRRLFy4sNIYXnzxRYYP34znn6/2amjUqMMrPxY2tLkPwMysppwAzMxqygnAzKymnADMzGrKCcDMrKZ8F5CZ2QCYP38+hx9+NIsWvVhpHCNHDi+8rBOAmdkAeOihh7jnnkdZvPicSuMYOfLkwss6AZiZDZA113wtsH/FMYwuvGxlfQCSDpI0U9L9kr5QVRxmZnVVSQKQtDbwI+BA4C3A4ZLeWkUsA6Grq6vqEArqqjqAgrqqDqBXr776QtUhFDJU/m06zmpUdQXwDmBWRDwZEa8ClwHvrSiWfhs6/yi6qg6goK6qA+jV0qXVdvQVNVT+bTrOalSVADYD5jS8n5vnmZlZhwyBTuBJFe77WV54YQGf+tSnelzqnnvu4cknnywtiokTfwK8h/7nyAep9ngW1T7OF1+cxOmnT+W2227rbEgN/vrXv7JkybNUfSwXLpzEhAkPMHny5LbLlP1vc6D8I8Q5e/ZsnnvuD1T97+KVV54uvKwiosRQ2uxU2gv4QkQckt//H2CdiPhG03KdD87M7B9ARPQ6UERVVwB3AttL2gSYD3wIOLZ5oSIfwMzMVk8lCSAiXpZ0HDAZEPDziJhWRSxmZnVVSROQmZlVb1AWg5N0vqR5kqodcqkHkjaTdEt+mO1Pkk6tOqZWJK0j6S5J0yQ9KOmsqmPqiaRhOdarq46lHUmPSbpP0r2S7qw6nnYkvUbS5TnW2ZJ2rTqmZpK2ycdxWv75nKQTqo6rmaSvSXpI0gOSrpC0btUxtSLpiznOmUWO46C8ApC0J7AImBgRY6uOpxVJGwMbRcQsSesB04DDI2LQJS1J60bEi5LWAG4DvhgRXRWH1ZKkk4CdgVERcWjV8bQi6RFg54hYUHUsPZF0OfCriLhU0jBgZEQM2jEkc4xzgXdExJzelu8USVsBvwXeHBFLJF0G3BgRF1Qc2kokvQ34GbAL8CpwA3BSRMxqt86gvAKIiKnAoP7PFRHzug9sRCwCZgCbVhtVaxHR/dTSOqS/+VMVhtOWpM2Ag4H/qTqWXohB+n+nm6QNgLdGxKUAEbFsMH/5Z+8C/jKYvvyzZ4ElwEhJawIjgCeqDamlNwN/jIiXI2IpcAvwvp5WGNT/iIcKSVuSsu7UaiNpLTer3Av8DeiKiNlVx9TG94BTgMF3WbqyZcDk3LTy2aqDaeNNwNO5CWiWpJ9JGll1UL34EHBJ1UE0y1d6Z5K+9P8K/D0ibqo2qpZmAntLGi1pBOlk6vU9reAE0E+5+ecK4MTBeoaVz/52Ij1J9k5Je1cdUzNJ7wXmRcR00hn2YL4FeLeI2Jl0xnq0pGrLP7Y2DBgHfDsidiBdUZ9WbUjtSVoLOJT0f2lQkfRG4CRgC2ATYD1JH6k2qlVFxEzgLNKZ/82kVokeT6acAPohXw5eCfwiIn5TdTy9iYjngWuBQdcZCOwBHJrb1y8B9pU0seKYWoqIp/LP+aS//7hqI2ppDjA3Iu7O768EBnPBxfcA9+RjOti8HbgtIp7NTSu/AvasOKaWIuK8iBgbEbsC/w/4U0/LD+YEMNjPAgEuAGZHxNlVB9KOpDH5KoV858IBQNtOoapExJcj4vUR8UbgSODmiPhE1XE1kzSi+w6Q3KRyEDDomtQiYi6pCehNedb+9PJlULEPMwibf7K/ALtKWleSSMfyLxXH1JKkMfnn64AjSIU22xqUtYAkXQzsA4yR9AQwISIurDaqlUnaA/goMDO3rwfw5Yi4odrIVrEJMDH9u2U4cHFEXFttSEPaxsBVkpaROgMvjYjBesvqvwEX54T1BOnf66CT26vfBfRcdKsiEXGXpCtJTSpLgenAD6qNqq2rJK0PvAJ8JiLm9bTwoLwN1MzMyjeYm4DMzKxETgBmZjXlBGBmVlNOAGZmNeUEYGZWU04AZmY15QRQU5KWSfpOw/vPS/rKAG37QknjB2JbvezniFz69nf93M6UXEmxr+u9RdJ7Gt6/r6yy4JK2kPTh1VjvKEnn5OljJX1sIOPI2z+3j9s4UdLw/sRhA8MJoL5eBsbnqpGDRi4JXNTRwCcjoqpaPG8lFdwCICImRcS3S9rXG4B+1Z+JiP+OiItKiKOvDxN9jvQQnVXMCaC+XgV+DJzc/IvmM3hJC/PPvSV1SfqlpD9L+k9JH5N0u9KgOFs3bOYASX+U9LCkD+T115B0rqTpeYCS4xu2e6ukX5MqGjbHc3Re/n5J38vzTiPVYzlf0realt9UabCeaZJm5PElus/Q78nzruoukdG0bstlJO0h6e4c+52SRgGnA0fk/Xyw+2xY0ihJjzVsc4SkJ/Ln3zZfcUzPx2e7FjHspxWDpEzLMXwT2DO/P7H5zFvSJEnvzNPHSvqLpNtINZa6l5kg6eQ8vU2rOCR9WGkwkWmSbm2OrTmOPG9TSdfkfS4viyLpvHysHpT0zTzveNLT6VP6e+VmAyAi/KrhC3geWA94FFgf+Dzwlfy7C4Hxjcvmn3uTaqOPAdYmlcY9Lf/uBODchvWvydNbAvNIZSiOJ5XLIK9/D7B13u5CYJMWcW5OGiTkNaTaUL8Fjsi/mwLs1GKdU4BTGt6PAP4J+AOwbp53KnBGw3be1m6ZHOtcYGyevy6wBnAUcE7Dfpa/B34N7J2njwB+nKenAlvl6bcDU1vEfy0wLk93j+GwN3B1q33l95OAd5Iqvs4FRuX1bm2IaQJwck9xAPeTBjqCNIBMc2yt4ng4H5N1gMeBLfPv1s8/h3Uf4/z+EWB01f8H/IrBWQvIOiMiFkn6GXAi8GJvy2d3RcQzAJIeBrrros8E9mtY7sq8j8ckPQDsALwbeJOkD+ZlRgFbAS8Bd0bEky32tytwU0Q8l/d5CbAXcHn+fauCgX8E/kepBs41ETEtt9W/CbhNkoC1gDua1turxTJ3AmOBRyOP9hZ5gJ20SFuXk+rb30IqbvcDpUJdOwNXaMXKrdrCbwXOyZ/11xExp5d9NdqNdLyezzFekT/Tcr3EcStwkaRfkqpeLi6wz981HJP7SUnoMeDfJH2S1ET0z6QBS6YxNAo91oITgH2f9J+ysdjeMnLzYP6CWLvhdy83Lfdy8zpZY7uwGt4fFxFTGgNQGp+gpy+aPn1ZRMTv8zYPBn4i6fvAC8B1EXFUL/tZZRlJu/Q1BuBq4BuSRpOuLm4mXTk9FRE9djhHxLckXZPjnyrpgBaLNR/v7i/wKBCr2sUREcdJGpf3fY+kt0bvQ182/ptYCgyTtA3wGdKoZIskXYi/bwYd9wHUl2D5aEeXA8c0/G4uaYQzgENIZ8J9dRgsHy1tG1IJ6snAscodvZK2VO+Da99OGhvgNXm97rPqtpSGlpwXaczWC/Jn+X3ezuvzMusojfXaqN0yM4AtJL0lzx+hNL7yi0DLUbYiYjFwNynBXhPJ08B8SYc0xNqqD2CLiLg/Ir5DugLZocW+5gJvVbIpqRkH0tXPvrkfYg3g8BaxtY0j7/uuiPgaaQS5NzSt3vYzNxlOatZbLGlDUr3/bi8U3IaVzAmgvhrP0M8knZ12zzsPOFDSNGB32p+d93T3x1xJt5OaiI6LiJdJJXT/Ctwv6T7gp/RyVhiprv1XSF9ss0jjL1zZy/73B2bk+I8Avh+pLO6ngKuVynffCXR/+UbeV8tlImIJKfFcKGk66Wx+HVK79s65I/WDrOoyUgnmSxvmfQj4vNJwkrOAVuudotQJfR+ps/5aUgnitfP8E4EuVgz4cTapP6X7eH0duJeUKB9oc4yObBPH2XneDNL4stOa1muOo/lv0H0sZ5D+Xg8BF7HycKkX4E7gQcHloM3MaspXAGZmNeUEYGZWU04AZmY15QRgpVIf6+woPRU8qeH9mpLuKSe65fu4RunJ3tVZt9daOPkz7bYa235UJZXqaD7OVk++L9cGo8Y7E/Zk5TtIBn5nEYf0vlTPm+jl9/sAi0i3tA7kdvvLd4DUnK8AakbS+pKuzbcuzpB0RJ4/QdIdkh5QqgWkPH+KpLOU6sXMljRO0pVKNX6+lZfZIq83UdIspbo0q9zfr/Z1dv6XUm2hPwLNVUQPAq5v2Mf5SnWHfiHpAKUaQo90n2Hn+G7Pn+9uSf+S5x+lVMOoVc2aRyVt0NPnaDwbl7SzpClNcXZ/vjvy57tF0uskbQH8b+BzSvVz9pC0cY5julLNn73z+hvmz3OvpB/T5oGuHo5ju7/hmyX9Pt/eOU1S973960u6VKmi6uVt9rVH3t4dkr4taWae31Ndpymttitp1/y3uU/SzZI2abVP66Cqa1H41dkX6X7vHzS8H5l/rt8wbyK5FhDpXvdv5OkTgCdZuRbQRsAWpCdTu+vX/Bj4YsP6PdXZGU66n32LPP9iVq41c0deZgtgCbBtnn838JM8fSgrag+NbFh3f2BSnu6pZs0jwAa9fI5HgA3y9M7AzQ3bPafFMTyGFjV48vtfArvn6c2Bh/P0ecCX8vS7SU/VbtD092t1HL/ey99wOnBwnl4jH8+9gQXAP+X5fyDXLmra34PAznn668CMPN1TXadVtkt6mPBuYEyefwRwUdX/H+r+chNQ/UwHvqlUnfH6iOiu+Pg+SZ8nNQtuQHrAqNs1+edMYGasXAtoU9J/+Cci4q683CWk4nKN2tXZ2RF4MCIeb1j33/P2NwGeiYiX8snsoxHxYF7uftIDWd1xbZanN5J0GakI3TJWrrXTrmZN45l2u89RpBTE1pLOAjYkHcc5bZZ7F/CG7jN00oNVo0jH6D8BImKypFYlGNodR2jxN1R6CndMRFyXt7sUWJp3fWdEPJXXnU5KRstJ2ghYOyK6+2AuIyVb6L2uU/N2x+a4f5vjHkYqEmgVcgKomYj4s1Jtm4OBr0m6GTiL9DTp2Ij4m6QJrPxvo7HeT3MtoHbNiM3ty6tTZ+cg4MYWcTTH0hjHN/J+fpibX6a0WX9pD7E36v4cjftoN5jJf5HOxq/PzToTetjmuPxlvJykZU3LtTou7Y7jSNr/Ddsd39U5Ho1xtKvr1Gq7AqZHxN592IeVzH0ANSPpdcDiiLgY+C4wjvRFsRRYkNu8V6kfU8Dr85c5pHIHzR237erszCKdSb6+Yd1uBwHXN4ZfII51SU1KAJ/oQ/zd2n2OOaSmH4APtFl3OKl+DsDHG+Y318+5CTiu+42k7fPkVFKJBpQKwL22xT7aHceWf8NIdX+eUq77o3RXVW/1l8jrzgeWaMVdXI1lK26kb3WdZpCO7U55+TUkbVskDiuPE0D9jCVVeZxGqrFzRqRSyz8ltfdez4omBej5TpHG3z0IHK9UV2Yz4JzGZaJ9nZ2XSJ2kv1PqBJ4Py0cG2zoiHmqzv3ZxfRc4U9Kd9FzErt222n2O04HzlOobNZ/k9izlAAAA2klEQVSpd/s6cFX+HM80zJ8EfCR37u5B+rwHKA28MpNUjhvS3+PgfHwOA55YJej2x7Gnv+FHgC8p1/cBNu7leDT6V+AXku4gNfN0lw0vWtep+++/hJRAzstxTyf1DViFXAvI+i03tVwTETsO4Db3AD4aEZ8eqG0W2OeAf46hTtK6Df0mXyB11nfsb2Llch+ADZQBPZOIiNuA2wZym0V3XcE+B7NDJX2J1LQ2h36OS2yDi68AzMxqyn0AZmY15QRgZlZTTgBmZjXlBGBmVlNOAGZmNeUEYGZWU/8fNtJQdSHeYdwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f30b7b2d400>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "temp = transformed_subset.s_source.apply(lambda x: len(x))\n",
    "plt.hist(temp, bins = 10, histtype = 'stepfilled')\n",
    "ax.set_xlabel('Number of selective studies that\\nsampled/manipulated each gene')\n",
    "ax.set_ylabel('Number of unique genes')\n",
    "plt.savefig('simulations_output/num_s_source_per_unique_gene.png', bbox_inches = \"tight\", dpi = 300) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
