{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate SFS_and_divergence.pickle\n",
    "\n",
    "- Use divergence data provided by Rob\n",
    "- Merge divergence data with `final_SFS.pickle` to generate `SFS_and_divergence.pickle`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notes from Rob\n",
    "in this file: `/scratch/research/projects/chlamydomonas/mutability/analysis/calculate_divergence/Cinc/K4K0.out.txt`\n",
    "you can find a table of divergence\n",
    "\n",
    "- `transcript_ID` = C. reinhardtii transcript ID \n",
    "- `Cincerta_transcript_ID` = C. incerta transcript ID \n",
    "- `aln_len` = length of aligned sequences\n",
    "- `Cr_length` = length of Creinhardtii transcript unaligned\n",
    "-  `Ci_length` = length of Cincerta transcript unaligned\n",
    "- `diff_percent` = proportion of sites that differed in the best BLAST hit between these sequences\n",
    "- `diffs0` = Number of differences at 0-fold degenerate sites\n",
    "- `sites0` = Number of 0-fold degenerate sites\n",
    "- `diffs4` = Number of differences at 4-fold degenerate sites\n",
    "- `sites4` = Number of 4-fold degenerate sites\n",
    "- `k0` = diffs0/sites0\n",
    "- `k4` = diffs4/sites4\n",
    "\n",
    "\n",
    "those are the columns in the table\n",
    "\n",
    "This file `/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/concatenated_GFF/ness_id.translation_table.txt` will convert from the transcript ID above to gene IDs. The first column of the translation table is “ness_id” - that equates to transcript_ID from the K0K4 table.\n",
    "\n",
    "- Use ness_id  (first column) and CDS (6th column) to convert to 5.3 gene ids\n",
    "- The diffs and sites columns are the data you need for DFE alpha\n",
    "\n",
    "**IMPORTANT CAVEAT** - some of the alignments are bound to be shit - you should probably remove any alignment where K4 > 0.6 or so"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num rows in divergence: 12294\n",
      "Num rows in divergence after filter: 12273\n",
      "Num rows removed in divergence due to filter: 21\n"
     ]
    }
   ],
   "source": [
    "#Import divergence and translation tables\n",
    "f = \"/scratch/research/projects/chlamydomonas/mutability/analysis/calculate_divergence/Cinc/K4K0.out.txt\"\n",
    "divergence = pd.read_csv(f, sep = \"\\t\")\n",
    "d1=len(divergence)\n",
    "print(\"Num rows in divergence:\", len(divergence))\n",
    "\n",
    "#Remove alignment where K4>0.6\n",
    "divergence = divergence[divergence.k4 <= 0.6]\n",
    "print(\"Num rows in divergence after filter:\", len(divergence))\n",
    "print(\"Num rows removed in divergence due to filter:\", d1-len(divergence))\n",
    "\n",
    "f = \"/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/concatenated_GFF/ness_id.translation_table.txt\"\n",
    "ness_id_table = pd.read_csv(f,sep = \"\\t\")\n",
    "ness_id_table.ness_id = ness_id_table.index\n",
    "ness_id_table['transcript_ID']= ness_id_table.index\n",
    "ness_id_table.index = range(len(ness_id_table))\n",
    "\n",
    "\n",
    "divergence.transcript_ID=divergence.transcript_ID.apply(lambda x: str(x))\n",
    "ness_id_table.transcript_ID=ness_id_table.transcript_ID.apply(lambda x: str(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>Cincerta_transcript_ID</th>\n",
       "      <th>aln_len</th>\n",
       "      <th>Cr_length</th>\n",
       "      <th>Ci_length</th>\n",
       "      <th>diff_percent</th>\n",
       "      <th>diffs0</th>\n",
       "      <th>sites0</th>\n",
       "      <th>diffs4</th>\n",
       "      <th>sites4</th>\n",
       "      <th>k0</th>\n",
       "      <th>k4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26888200</td>\n",
       "      <td>g6935.t1</td>\n",
       "      <td>4230</td>\n",
       "      <td>4230</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.05</td>\n",
       "      <td>24</td>\n",
       "      <td>2777</td>\n",
       "      <td>113</td>\n",
       "      <td>699</td>\n",
       "      <td>0.008642</td>\n",
       "      <td>0.161660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26904954</td>\n",
       "      <td>g13204.t1</td>\n",
       "      <td>1041</td>\n",
       "      <td>1041</td>\n",
       "      <td>1032</td>\n",
       "      <td>0.09</td>\n",
       "      <td>28</td>\n",
       "      <td>655</td>\n",
       "      <td>38</td>\n",
       "      <td>217</td>\n",
       "      <td>0.042748</td>\n",
       "      <td>0.175115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26894252</td>\n",
       "      <td>g11333.t1</td>\n",
       "      <td>4617</td>\n",
       "      <td>4323</td>\n",
       "      <td>4521</td>\n",
       "      <td>0.15</td>\n",
       "      <td>231</td>\n",
       "      <td>2638</td>\n",
       "      <td>311</td>\n",
       "      <td>950</td>\n",
       "      <td>0.087566</td>\n",
       "      <td>0.327368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26891814</td>\n",
       "      <td>g15729.t1</td>\n",
       "      <td>5127</td>\n",
       "      <td>4308</td>\n",
       "      <td>4932</td>\n",
       "      <td>0.22</td>\n",
       "      <td>450</td>\n",
       "      <td>2626</td>\n",
       "      <td>325</td>\n",
       "      <td>930</td>\n",
       "      <td>0.171363</td>\n",
       "      <td>0.349462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26901315</td>\n",
       "      <td>g5327.t1</td>\n",
       "      <td>2289</td>\n",
       "      <td>2289</td>\n",
       "      <td>2289</td>\n",
       "      <td>0.05</td>\n",
       "      <td>3</td>\n",
       "      <td>1483</td>\n",
       "      <td>72</td>\n",
       "      <td>412</td>\n",
       "      <td>0.002023</td>\n",
       "      <td>0.174757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_ID Cincerta_transcript_ID  aln_len  Cr_length  Ci_length  \\\n",
       "0      26888200               g6935.t1     4230       4230       4230   \n",
       "1      26904954              g13204.t1     1041       1041       1032   \n",
       "2      26894252              g11333.t1     4617       4323       4521   \n",
       "3      26891814              g15729.t1     5127       4308       4932   \n",
       "4      26901315               g5327.t1     2289       2289       2289   \n",
       "\n",
       "   diff_percent  diffs0  sites0  diffs4  sites4        k0        k4  \n",
       "0          0.05      24    2777     113     699  0.008642  0.161660  \n",
       "1          0.09      28     655      38     217  0.042748  0.175115  \n",
       "2          0.15     231    2638     311     950  0.087566  0.327368  \n",
       "3          0.22     450    2626     325     930  0.171363  0.349462  \n",
       "4          0.05       3    1483      72     412  0.002023  0.174757  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "divergence[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ness_id</th>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <th>three_prime_UTR</th>\n",
       "      <th>exon</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>CDS</th>\n",
       "      <th>tRNA</th>\n",
       "      <th>rRNA</th>\n",
       "      <th>misc_RNA</th>\n",
       "      <th>transcript_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26903746</td>\n",
       "      <td>PAC:26903746</td>\n",
       "      <td>PAC:26903746</td>\n",
       "      <td>PAC:26903746</td>\n",
       "      <td>g2</td>\n",
       "      <td>PAC:26903746</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>26903746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26903463</td>\n",
       "      <td>PAC:26903463</td>\n",
       "      <td>PAC:26903463</td>\n",
       "      <td>PAC:26903463</td>\n",
       "      <td>g3</td>\n",
       "      <td>PAC:26903463</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>26903463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26903339</td>\n",
       "      <td>PAC:26903339</td>\n",
       "      <td>PAC:26903339</td>\n",
       "      <td>PAC:26903339</td>\n",
       "      <td>Cre01.g000050</td>\n",
       "      <td>PAC:26903339</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26903974</td>\n",
       "      <td>PAC:26903974</td>\n",
       "      <td>PAC:26903974</td>\n",
       "      <td>PAC:26903974</td>\n",
       "      <td>Cre01.g000100</td>\n",
       "      <td>PAC:26903974</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>26903974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26903809</td>\n",
       "      <td>PAC:26903809</td>\n",
       "      <td>PAC:26903809</td>\n",
       "      <td>PAC:26903809</td>\n",
       "      <td>Cre01.g000150</td>\n",
       "      <td>PAC:26903809</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>26903809</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ness_id five_prime_UTR three_prime_UTR          exon           mRNA  \\\n",
       "0  26903746   PAC:26903746    PAC:26903746  PAC:26903746             g2   \n",
       "1  26903463   PAC:26903463    PAC:26903463  PAC:26903463             g3   \n",
       "2  26903339   PAC:26903339    PAC:26903339  PAC:26903339  Cre01.g000050   \n",
       "3  26903974   PAC:26903974    PAC:26903974  PAC:26903974  Cre01.g000100   \n",
       "4  26903809   PAC:26903809    PAC:26903809  PAC:26903809  Cre01.g000150   \n",
       "\n",
       "            CDS tRNA rRNA misc_RNA transcript_ID  \n",
       "0  PAC:26903746    .    .        .      26903746  \n",
       "1  PAC:26903463    .    .        .      26903463  \n",
       "2  PAC:26903339    .    .        .      26903339  \n",
       "3  PAC:26903974    .    .        .      26903974  \n",
       "4  PAC:26903809    .    .        .      26903809  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ness_id_table[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.merge(divergence, ness_id_table[['exon', 'mRNA','transcript_ID']], on = 'transcript_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All transcripts in divergence has matching PAC id in ness_id_table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>Cincerta_transcript_ID</th>\n",
       "      <th>aln_len</th>\n",
       "      <th>Cr_length</th>\n",
       "      <th>Ci_length</th>\n",
       "      <th>diff_percent</th>\n",
       "      <th>diffs0</th>\n",
       "      <th>sites0</th>\n",
       "      <th>diffs4</th>\n",
       "      <th>sites4</th>\n",
       "      <th>k0</th>\n",
       "      <th>k4</th>\n",
       "      <th>exon</th>\n",
       "      <th>mRNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26888200</td>\n",
       "      <td>g6935.t1</td>\n",
       "      <td>4230</td>\n",
       "      <td>4230</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.05</td>\n",
       "      <td>24</td>\n",
       "      <td>2777</td>\n",
       "      <td>113</td>\n",
       "      <td>699</td>\n",
       "      <td>0.008642</td>\n",
       "      <td>0.161660</td>\n",
       "      <td>PAC:26888200</td>\n",
       "      <td>Cre08.g362650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26904954</td>\n",
       "      <td>g13204.t1</td>\n",
       "      <td>1041</td>\n",
       "      <td>1041</td>\n",
       "      <td>1032</td>\n",
       "      <td>0.09</td>\n",
       "      <td>28</td>\n",
       "      <td>655</td>\n",
       "      <td>38</td>\n",
       "      <td>217</td>\n",
       "      <td>0.042748</td>\n",
       "      <td>0.175115</td>\n",
       "      <td>PAC:26904954</td>\n",
       "      <td>Cre10.g442950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26894252</td>\n",
       "      <td>g11333.t1</td>\n",
       "      <td>4617</td>\n",
       "      <td>4323</td>\n",
       "      <td>4521</td>\n",
       "      <td>0.15</td>\n",
       "      <td>231</td>\n",
       "      <td>2638</td>\n",
       "      <td>311</td>\n",
       "      <td>950</td>\n",
       "      <td>0.087566</td>\n",
       "      <td>0.327368</td>\n",
       "      <td>PAC:26894252</td>\n",
       "      <td>Cre06.g311400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26891814</td>\n",
       "      <td>g15729.t1</td>\n",
       "      <td>5127</td>\n",
       "      <td>4308</td>\n",
       "      <td>4932</td>\n",
       "      <td>0.22</td>\n",
       "      <td>450</td>\n",
       "      <td>2626</td>\n",
       "      <td>325</td>\n",
       "      <td>930</td>\n",
       "      <td>0.171363</td>\n",
       "      <td>0.349462</td>\n",
       "      <td>PAC:26891814</td>\n",
       "      <td>Cre16.g674500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26901315</td>\n",
       "      <td>g5327.t1</td>\n",
       "      <td>2289</td>\n",
       "      <td>2289</td>\n",
       "      <td>2289</td>\n",
       "      <td>0.05</td>\n",
       "      <td>3</td>\n",
       "      <td>1483</td>\n",
       "      <td>72</td>\n",
       "      <td>412</td>\n",
       "      <td>0.002023</td>\n",
       "      <td>0.174757</td>\n",
       "      <td>PAC:26901315</td>\n",
       "      <td>Cre03.g180850</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_ID Cincerta_transcript_ID  aln_len  Cr_length  Ci_length  \\\n",
       "0      26888200               g6935.t1     4230       4230       4230   \n",
       "1      26904954              g13204.t1     1041       1041       1032   \n",
       "2      26894252              g11333.t1     4617       4323       4521   \n",
       "3      26891814              g15729.t1     5127       4308       4932   \n",
       "4      26901315               g5327.t1     2289       2289       2289   \n",
       "\n",
       "   diff_percent  diffs0  sites0  diffs4  sites4        k0        k4  \\\n",
       "0          0.05      24    2777     113     699  0.008642  0.161660   \n",
       "1          0.09      28     655      38     217  0.042748  0.175115   \n",
       "2          0.15     231    2638     311     950  0.087566  0.327368   \n",
       "3          0.22     450    2626     325     930  0.171363  0.349462   \n",
       "4          0.05       3    1483      72     412  0.002023  0.174757   \n",
       "\n",
       "           exon           mRNA  \n",
       "0  PAC:26888200  Cre08.g362650  \n",
       "1  PAC:26904954  Cre10.g442950  \n",
       "2  PAC:26894252  Cre06.g311400  \n",
       "3  PAC:26891814  Cre16.g674500  \n",
       "4  PAC:26901315  Cre03.g180850  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if len(temp.dropna()) == len(divergence): print(\"All transcripts in divergence has matching PAC id in ness_id_table\")\n",
    "temp[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAC_id</th>\n",
       "      <th>neutral_SFS</th>\n",
       "      <th>selected_SFS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAC:26887927</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAC:26887929</td>\n",
       "      <td>[126, 9, 0, 1, 1, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>[399, 9, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PAC:26887930</td>\n",
       "      <td>[273, 12, 11, 3, 5, 6, 6, 3, 8, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>[921, 22, 15, 9, 3, 6, 1, 1, 10, 0, 0, 0, 0, 0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         PAC_id                                        neutral_SFS  \\\n",
       "0  PAC:26887927  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "1  PAC:26887929  [126, 9, 0, 1, 1, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0...   \n",
       "2  PAC:26887930  [273, 12, 11, 3, 5, 6, 6, 3, 8, 0, 0, 0, 0, 0,...   \n",
       "\n",
       "                                        selected_SFS  \n",
       "0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "1  [399, 9, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0...  \n",
       "2  [921, 22, 15, 9, 3, 6, 1, 1, 10, 0, 0, 0, 0, 0...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reduced_SFS = pickle.load(open('../../data/intermediate_data_02/reduced_SFS.pk', 'rb'))\n",
    "reduced_SFS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cincerta_transcript_ID</th>\n",
       "      <th>diffs0</th>\n",
       "      <th>sites0</th>\n",
       "      <th>diffs4</th>\n",
       "      <th>sites4</th>\n",
       "      <th>PAC_id</th>\n",
       "      <th>neutral_SFS</th>\n",
       "      <th>selected_SFS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g6935.t1</td>\n",
       "      <td>24</td>\n",
       "      <td>2777</td>\n",
       "      <td>113</td>\n",
       "      <td>699</td>\n",
       "      <td>PAC:26888200</td>\n",
       "      <td>[609, 11, 17, 29, 6, 10, 6, 2, 8, 0, 0, 0, 0, ...</td>\n",
       "      <td>[2775, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g13204.t1</td>\n",
       "      <td>28</td>\n",
       "      <td>655</td>\n",
       "      <td>38</td>\n",
       "      <td>217</td>\n",
       "      <td>PAC:26904954</td>\n",
       "      <td>[155, 4, 0, 2, 0, 4, 3, 3, 1, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>[519, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g11333.t1</td>\n",
       "      <td>231</td>\n",
       "      <td>2638</td>\n",
       "      <td>311</td>\n",
       "      <td>950</td>\n",
       "      <td>PAC:26894252</td>\n",
       "      <td>[491, 4, 9, 4, 1, 0, 7, 2, 4, 1, 0, 0, 0, 0, 0...</td>\n",
       "      <td>[1435, 9, 2, 1, 0, 0, 4, 2, 3, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cincerta_transcript_ID  diffs0  sites0  diffs4  sites4        PAC_id  \\\n",
       "0               g6935.t1      24    2777     113     699  PAC:26888200   \n",
       "1              g13204.t1      28     655      38     217  PAC:26904954   \n",
       "2              g11333.t1     231    2638     311     950  PAC:26894252   \n",
       "\n",
       "                                         neutral_SFS  \\\n",
       "0  [609, 11, 17, 29, 6, 10, 6, 2, 8, 0, 0, 0, 0, ...   \n",
       "1  [155, 4, 0, 2, 0, 4, 3, 3, 1, 0, 0, 0, 0, 0, 0...   \n",
       "2  [491, 4, 9, 4, 1, 0, 7, 2, 4, 1, 0, 0, 0, 0, 0...   \n",
       "\n",
       "                                        selected_SFS  \n",
       "0  [2775, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, ...  \n",
       "1  [519, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0...  \n",
       "2  [1435, 9, 2, 1, 0, 0, 4, 2, 3, 0, 0, 0, 0, 0, ...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = temp.rename(columns={'exon':'PAC_id'})\n",
    "output = pd.merge(temp[['Cincerta_transcript_ID','diffs0','sites0','diffs4','sites4','PAC_id']], reduced_SFS, on = 'PAC_id')\n",
    "output[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/workflow/SFS_and_divergence.pk', 'wb') as f:\n",
    "    pickle.dump(output, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
