{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract candidate gene info from *Chlamy* gff\n",
    "\n",
    "gff file path: `scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/Creinhardtii_v5.3_223_gene.gff3.gz `\n",
    "\n",
    "Target data structure:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import re\n",
    "import time\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.read_csv(\"../../data/intermediate_data_from_gff/merged_v5_3_1.csv\")\n",
    "gff_file_path='/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/Creinhardtii_v5.3_223_gene.gff3.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BinarySearch(lys, val):\n",
    "    \n",
    "    '''requires re'''\n",
    "    \n",
    "    '''This function returns the position of the element in the list lys that contains the string pattern lys. If no match \n",
    "    Usage: lys = list of strings to search through; val = string pattern to search for\n",
    "    Warning: This function only works when the beginning of the string matches val'''\n",
    "        \n",
    "    first = 0\n",
    "    last = len(lys)-1\n",
    "    index = -1\n",
    "    \n",
    "    p = re.compile(re.escape(val))\n",
    "    \n",
    "    while (first <= last) and (index == -1):\n",
    "        \n",
    "        mid = round((first+last)/2)\n",
    "        index_list = sorted([lys[mid],val])\n",
    "        \n",
    "        if p.match(lys[mid]): index = mid\n",
    "        else:\n",
    "            \n",
    "            if index_list.index(val)<index_list.index(lys[mid]): last = mid -1\n",
    "            else: first = mid +1\n",
    "                \n",
    "    return index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tryextract(i, pattern):\n",
    "    \n",
    "    '''This function takes in a string and returns the first matching group in the search pattern'''\n",
    "    \n",
    "    try: \n",
    "        m = re.search(pattern, i).group(1)\n",
    "        return(m)\n",
    "    \n",
    "    except AttributeError:\n",
    "        return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import gff as working dataframe. Extract Name and PAC id from the column 'attribute' as separate columns\n",
    "with gzip.open(gff_file_path, \"rt\", encoding=\"utf-8\") as z:\n",
    "    \n",
    "    df = pd.read_csv(z,delimiter=r\"\\s+\",skiprows=1,header=None)\n",
    "    df.columns = ['chromosome', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    df['Name'] = df.attributes.apply(lambda x: tryextract(x, r\"Name=(.+);pacid\"))\n",
    "    df['ID'] = df.attributes.apply(lambda x: tryextract(x, r\"ID=(PAC:[0-9]+)\"))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create mRNA gff dataframe subset to extract PAC_id \n",
    "subset = df[df.feature ==\"mRNA\"].sort_values(by=\"Name\").reset_index()\n",
    "search_list = list(subset.Name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   index    chromosome        source feature  start    end score strand phase  \\\n",
      "0      1  chromosome_1  phytozome8_0    mRNA  24026  30617     .      +     .   \n",
      "1      9  chromosome_1  phytozome8_0    mRNA  30776  41037     .      +     .   \n",
      "2     26  chromosome_1  phytozome8_0    mRNA  41896  46547     .      +     .   \n",
      "3     35  chromosome_1  phytozome8_0    mRNA  46553  51525     .      -     .   \n",
      "4     51  chromosome_1  phytozome8_0    mRNA  51590  54035     .      -     .   \n",
      "5     59  chromosome_1  phytozome8_0    mRNA  54136  57919     .      +     .   \n",
      "6     71  chromosome_1  phytozome8_0    mRNA  57977  60685     .      -     .   \n",
      "7     81  chromosome_1  phytozome8_0    mRNA  60740  64955     .      -     .   \n",
      "8     96  chromosome_1  phytozome8_0    mRNA  65056  70197     .      +     .   \n",
      "9    109  chromosome_1  phytozome8_0    mRNA  70246  77235     .      +     .   \n",
      "\n",
      "                                          attributes                Name  \\\n",
      "0  ID=PAC:26903339;Name=Cre01.g000050.t1.3;pacid=...  Cre01.g000050.t1.3   \n",
      "1  ID=PAC:26903974;Name=Cre01.g000100.t1.3;pacid=...  Cre01.g000100.t1.3   \n",
      "2  ID=PAC:26903809;Name=Cre01.g000150.t1.2;pacid=...  Cre01.g000150.t1.2   \n",
      "3  ID=PAC:26903011;Name=Cre01.g000200.t1.3;pacid=...  Cre01.g000200.t1.3   \n",
      "4  ID=PAC:26903028;Name=Cre01.g000250.t1.2;pacid=...  Cre01.g000250.t1.2   \n",
      "5  ID=PAC:26903629;Name=Cre01.g000300.t1.3;pacid=...  Cre01.g000300.t1.3   \n",
      "6  ID=PAC:26903061;Name=Cre01.g000350.t1.3;pacid=...  Cre01.g000350.t1.3   \n",
      "7  ID=PAC:26903690;Name=Cre01.g000400.t1.2;pacid=...  Cre01.g000400.t1.2   \n",
      "8  ID=PAC:26903197;Name=Cre01.g000450.t1.2;pacid=...  Cre01.g000450.t1.2   \n",
      "9  ID=PAC:26902937;Name=Cre01.g000500.t1.3;pacid=...  Cre01.g000500.t1.3   \n",
      "\n",
      "             ID  \n",
      "0  PAC:26903339  \n",
      "1  PAC:26903974  \n",
      "2  PAC:26903809  \n",
      "3  PAC:26903011  \n",
      "4  PAC:26903028  \n",
      "5  PAC:26903629  \n",
      "6  PAC:26903061  \n",
      "7  PAC:26903690  \n",
      "8  PAC:26903197  \n",
      "9  PAC:26902937  \n"
     ]
    }
   ],
   "source": [
    "print(subset[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken 10.309276580810547 s\n"
     ]
    }
   ],
   "source": [
    "#Order list of matching PAC id's to 'transcript_id_v5.3.1'\n",
    "t0 = time.time()\n",
    "pacid_list = []\n",
    "\n",
    "for row in range(0,len(merged)):\n",
    "    transcript_id = merged.loc[row,'transcript_id_v5.3.1']\n",
    "    pacid = None\n",
    "    \n",
    "    index = BinarySearch(search_list, transcript_id)\n",
    "    \n",
    "    if index > -1: \n",
    "        pacid = subset.loc[index, \"ID\"]\n",
    "                    \n",
    "    pacid_list.append(pacid)\n",
    "\n",
    "t1 = time.time()\n",
    "print(\"Time taken\", t1-t0, \"s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add column of PAC id's associated with each 'transcript_id_v5.3.1' in merged dataframe\n",
    "merged['PAC_ID'] = pacid_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define candidate gene conditions\n",
    "A = (merged.num_detected>=2)\n",
    "B = (merged.num_sampled>=2)\n",
    "C = merged.pathway_id.str.contains(\"TRIGLSYN-PWY\")==True\n",
    "D = merged.pathway_id.str.contains(\"PWY-4381\")==True\n",
    "F = (merged.num_detected>=1)\n",
    "\n",
    "condition = (A & (C | D))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set= []\n",
    "features_condition = ['five_prime_UTR', 'CDS', 'three_prime_UTR']\n",
    "#keep gff rows that contain CDS positions, remove gff rows that are missing PAC_ID's\n",
    "filtered_df = df[df.feature.isin(features_condition)].dropna(subset = ['ID'])\n",
    "\n",
    "#loop through sampled transcipts with PAC_ID's; transcript filtered by condition variable\n",
    "for transcript in list(merged[condition].PAC_ID): \n",
    "    \n",
    "    temp = filtered_df[filtered_df.ID==transcript]\n",
    "    \n",
    "    #addd CDS coordinates as tuple (chromosome, start, end) to gene_set\n",
    "    for i in list(temp.index):\n",
    "        gene_set.append((temp.chromosome[i], temp.start[i], temp.end[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/intermediate_data_from_gff/gene_set.pickle', 'wb') as f:\n",
    "    pickle.dump(gene_set, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "176 transcripts out of 17956 transcripts did not have a matching PAC_ID\n"
     ]
    }
   ],
   "source": [
    "print(list(merged['PAC_ID']).count(None),\"transcripts out of\", len(merged), \"transcripts did not have a matching PAC_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_detected</th>\n",
       "      <th>num_sampled</th>\n",
       "      <th>proportion</th>\n",
       "      <th>source</th>\n",
       "      <th>annotation_version</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>transcript_id_v5.3.1</th>\n",
       "      <th>PAC_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cre01.g000017.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016', 'Kwak_2017']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g2.t1</td>\n",
       "      <td>PAC:26903746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre01.g000033.t1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g3.t1</td>\n",
       "      <td>PAC:26903463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre01.g000050.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Gargouri_2015']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000050.t1.3</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cre01.g000100.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000100.t1.3</td>\n",
       "      <td>PAC:26903974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cre01.g000150.t1.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000150</td>\n",
       "      <td>ZRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000150.t1.2</td>\n",
       "      <td>PAC:26903809</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  num_detected  num_sampled  proportion  \\\n",
       "0  Cre01.g000017.t1.1             0            2       0.000   \n",
       "1  Cre01.g000033.t1.1             1            1       0.125   \n",
       "2  Cre01.g000050.t1.1             0            1       0.000   \n",
       "3  Cre01.g000100.t1.1             0            1       0.000   \n",
       "4  Cre01.g000150.t1.2             1            1       0.125   \n",
       "\n",
       "                           source annotation_version        gene_id  \\\n",
       "0  ['Bajhaiya_2016', 'Kwak_2017']               v5.5  Cre01.g000017   \n",
       "1               ['Bajhaiya_2016']               v5.5  Cre01.g000033   \n",
       "2               ['Gargouri_2015']               v5.5  Cre01.g000050   \n",
       "3               ['Bajhaiya_2016']               v5.5  Cre01.g000100   \n",
       "4               ['Bajhaiya_2016']               v5.5  Cre01.g000150   \n",
       "\n",
       "  gene_symbol pathway_id transcript_id_v5.3.1        PAC_ID  \n",
       "0         NaN        NaN                g2.t1  PAC:26903746  \n",
       "1         NaN        NaN                g3.t1  PAC:26903463  \n",
       "2         NaN        NaN   Cre01.g000050.t1.3  PAC:26903339  \n",
       "3         NaN        NaN   Cre01.g000100.t1.3  PAC:26903974  \n",
       "4        ZRT2        NaN   Cre01.g000150.t1.2  PAC:26903809  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>gene</td>\n",
       "      <td>24026</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Cre01.g000050;Name=Cre01.g000050</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24026</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339;Name=Cre01.g000050.t1.3;pacid=...</td>\n",
       "      <td>Cre01.g000050.t1.3</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>24026</td>\n",
       "      <td>24125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339.five_prime_UTR.1;Parent=PAC:26...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>24126</td>\n",
       "      <td>28105</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=PAC:26903339.CDS.1;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28291</td>\n",
       "      <td>28644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=PAC:26903339.CDS.2;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28842</td>\n",
       "      <td>29091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=PAC:26903339.CDS.3;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>29347</td>\n",
       "      <td>29577</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=PAC:26903339.CDS.4;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>29578</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339.three_prime_UTR.1;Parent=PAC:2...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>gene</td>\n",
       "      <td>30776</td>\n",
       "      <td>41037</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Cre01.g000100;Name=Cre01.g000100</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>30776</td>\n",
       "      <td>41037</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903974;Name=Cre01.g000100.t1.3;pacid=...</td>\n",
       "      <td>Cre01.g000100.t1.3</td>\n",
       "      <td>PAC:26903974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome        source          feature  start    end score strand  \\\n",
       "0  chromosome_1  phytozome8_0             gene  24026  30617     .      +   \n",
       "1  chromosome_1  phytozome8_0             mRNA  24026  30617     .      +   \n",
       "2  chromosome_1  phytozome8_0   five_prime_UTR  24026  24125     .      +   \n",
       "3  chromosome_1  phytozome8_0              CDS  24126  28105     .      +   \n",
       "4  chromosome_1  phytozome8_0              CDS  28291  28644     .      +   \n",
       "5  chromosome_1  phytozome8_0              CDS  28842  29091     .      +   \n",
       "6  chromosome_1  phytozome8_0              CDS  29347  29577     .      +   \n",
       "7  chromosome_1  phytozome8_0  three_prime_UTR  29578  30617     .      +   \n",
       "8  chromosome_1  phytozome8_0             gene  30776  41037     .      +   \n",
       "9  chromosome_1  phytozome8_0             mRNA  30776  41037     .      +   \n",
       "\n",
       "  phase                                         attributes  \\\n",
       "0     .                ID=Cre01.g000050;Name=Cre01.g000050   \n",
       "1     .  ID=PAC:26903339;Name=Cre01.g000050.t1.3;pacid=...   \n",
       "2     .  ID=PAC:26903339.five_prime_UTR.1;Parent=PAC:26...   \n",
       "3     0  ID=PAC:26903339.CDS.1;Parent=PAC:26903339;paci...   \n",
       "4     1  ID=PAC:26903339.CDS.2;Parent=PAC:26903339;paci...   \n",
       "5     1  ID=PAC:26903339.CDS.3;Parent=PAC:26903339;paci...   \n",
       "6     0  ID=PAC:26903339.CDS.4;Parent=PAC:26903339;paci...   \n",
       "7     .  ID=PAC:26903339.three_prime_UTR.1;Parent=PAC:2...   \n",
       "8     .                ID=Cre01.g000100;Name=Cre01.g000100   \n",
       "9     .  ID=PAC:26903974;Name=Cre01.g000100.t1.3;pacid=...   \n",
       "\n",
       "                 Name            ID  \n",
       "0                None          None  \n",
       "1  Cre01.g000050.t1.3  PAC:26903339  \n",
       "2                None  PAC:26903339  \n",
       "3                None  PAC:26903339  \n",
       "4                None  PAC:26903339  \n",
       "5                None  PAC:26903339  \n",
       "6                None  PAC:26903339  \n",
       "7                None  PAC:26903339  \n",
       "8                None          None  \n",
       "9  Cre01.g000100.t1.3  PAC:26903974  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3         ID=PAC:26903339.CDS.1;Parent=PAC:26903339;paci...\n",
      "4         ID=PAC:26903339.CDS.2;Parent=PAC:26903339;paci...\n",
      "5         ID=PAC:26903339.CDS.3;Parent=PAC:26903339;paci...\n",
      "6         ID=PAC:26903339.CDS.4;Parent=PAC:26903339;paci...\n",
      "11        ID=PAC:26903974.CDS.1;Parent=PAC:26903974;paci...\n",
      "12        ID=PAC:26903974.CDS.2;Parent=PAC:26903974;paci...\n",
      "13        ID=PAC:26903974.CDS.3;Parent=PAC:26903974;paci...\n",
      "14        ID=PAC:26903974.CDS.4;Parent=PAC:26903974;paci...\n",
      "15        ID=PAC:26903974.CDS.5;Parent=PAC:26903974;paci...\n",
      "16        ID=PAC:26903974.CDS.6;Parent=PAC:26903974;paci...\n",
      "17        ID=PAC:26903974.CDS.7;Parent=PAC:26903974;paci...\n",
      "18        ID=PAC:26903974.CDS.8;Parent=PAC:26903974;paci...\n",
      "19        ID=PAC:26903974.CDS.9;Parent=PAC:26903974;paci...\n",
      "20        ID=PAC:26903974.CDS.10;Parent=PAC:26903974;pac...\n",
      "21        ID=PAC:26903974.CDS.11;Parent=PAC:26903974;pac...\n",
      "22        ID=PAC:26903974.CDS.12;Parent=PAC:26903974;pac...\n",
      "23        ID=PAC:26903974.CDS.13;Parent=PAC:26903974;pac...\n",
      "28        ID=PAC:26903809.CDS.1;Parent=PAC:26903809;paci...\n",
      "29        ID=PAC:26903809.CDS.2;Parent=PAC:26903809;paci...\n",
      "30        ID=PAC:26903809.CDS.3;Parent=PAC:26903809;paci...\n",
      "31        ID=PAC:26903809.CDS.4;Parent=PAC:26903809;paci...\n",
      "32        ID=PAC:26903809.CDS.5;Parent=PAC:26903809;paci...\n",
      "36        ID=PAC:26903011.CDS.1;Parent=PAC:26903011;paci...\n",
      "38        ID=PAC:26903011.CDS.2;Parent=PAC:26903011;paci...\n",
      "39        ID=PAC:26903011.CDS.3;Parent=PAC:26903011;paci...\n",
      "40        ID=PAC:26903011.CDS.4;Parent=PAC:26903011;paci...\n",
      "41        ID=PAC:26903011.CDS.5;Parent=PAC:26903011;paci...\n",
      "42        ID=PAC:26903011.CDS.6;Parent=PAC:26903011;paci...\n",
      "43        ID=PAC:26903011.CDS.7;Parent=PAC:26903011;paci...\n",
      "44        ID=PAC:26903011.CDS.8;Parent=PAC:26903011;paci...\n",
      "                                ...                        \n",
      "250097    ID=PAC:26892883.CDS.7;Parent=PAC:26892883;paci...\n",
      "250098    ID=PAC:26892883.CDS.8;Parent=PAC:26892883;paci...\n",
      "250099    ID=PAC:26892883.CDS.9;Parent=PAC:26892883;paci...\n",
      "250100    ID=PAC:26892883.CDS.10;Parent=PAC:26892883;pac...\n",
      "250101    ID=PAC:26892883.CDS.11;Parent=PAC:26892883;pac...\n",
      "250102    ID=PAC:26892883.CDS.12;Parent=PAC:26892883;pac...\n",
      "250103    ID=PAC:26892883.CDS.13;Parent=PAC:26892883;pac...\n",
      "250107    ID=PAC:26892882.CDS.1;Parent=PAC:26892882;paci...\n",
      "250108    ID=PAC:26892882.CDS.2;Parent=PAC:26892882;paci...\n",
      "250110    ID=PAC:26892882.CDS.3;Parent=PAC:26892882;paci...\n",
      "250113    ID=PAC:26901101.CDS.1;Parent=PAC:26901101;paci...\n",
      "250115    ID=PAC:26901101.CDS.2;Parent=PAC:26901101;paci...\n",
      "250116    ID=PAC:26901101.CDS.3;Parent=PAC:26901101;paci...\n",
      "250117    ID=PAC:26901101.CDS.4;Parent=PAC:26901101;paci...\n",
      "250121    ID=PAC:26901100.CDS.1;Parent=PAC:26901100;paci...\n",
      "250122    ID=PAC:26901100.CDS.2;Parent=PAC:26901100;paci...\n",
      "250123    ID=PAC:26901100.CDS.3;Parent=PAC:26901100;paci...\n",
      "250124    ID=PAC:26901100.CDS.4;Parent=PAC:26901100;paci...\n",
      "250125    ID=PAC:26901100.CDS.5;Parent=PAC:26901100;paci...\n",
      "250126    ID=PAC:26901100.CDS.6;Parent=PAC:26901100;paci...\n",
      "250127    ID=PAC:26901100.CDS.7;Parent=PAC:26901100;paci...\n",
      "250128    ID=PAC:26901100.CDS.8;Parent=PAC:26901100;paci...\n",
      "250129    ID=PAC:26901100.CDS.9;Parent=PAC:26901100;paci...\n",
      "250130    ID=PAC:26901100.CDS.10;Parent=PAC:26901100;pac...\n",
      "250134    ID=PAC:26897502.CDS.1;Parent=PAC:26897502;paci...\n",
      "250135    ID=PAC:26897502.CDS.2;Parent=PAC:26897502;paci...\n",
      "250136    ID=PAC:26897502.CDS.3;Parent=PAC:26897502;paci...\n",
      "250137    ID=PAC:26897502.CDS.4;Parent=PAC:26897502;paci...\n",
      "250138    ID=PAC:26897502.CDS.5;Parent=PAC:26897502;paci...\n",
      "250139    ID=PAC:26897502.CDS.6;Parent=PAC:26897502;paci...\n",
      "Name: attributes, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(filtered_df.attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
