{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract candidate gene info from *Chlamy* gff\n",
    "\n",
    "gff file path: `scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/Creinhardtii_v5.3_223_gene.gff3.gz `\n",
    "\n",
    "Target data structure:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import re\n",
    "import time\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.read_csv(\"../../data/intermediate_data_from_gff/merged_v5_3_1.csv\")\n",
    "gff_file_path='/scratch/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/Creinhardtii_v5.3_223_gene.gff3.gz'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BinarySearch(lys, val):\n",
    "    \n",
    "    '''requires re'''\n",
    "    \n",
    "    '''This function returns the position of the element in the list lys that contains the string pattern lys. If no match \n",
    "    \n",
    "    Usage: lys = list of strings to search through; val = string pattern to search for\n",
    "    \n",
    "    Warning: This function only works when the beginning of the string matches val'''\n",
    "        \n",
    "    first = 0\n",
    "    last = len(lys)-1\n",
    "    index = -1\n",
    "    \n",
    "    \n",
    "    p = re.compile(re.escape(val))\n",
    "    \n",
    "    \n",
    "    while (first <= last) and (index == -1):\n",
    "        mid = round((first+last)/2)\n",
    "        \n",
    "        index_list = sorted([lys[mid],val])\n",
    "        \n",
    "        if p.match(lys[mid]):\n",
    "            index = mid\n",
    "        else:\n",
    "            \n",
    "            if index_list.index(val)<index_list.index(lys[mid]):\n",
    "                last = mid -1\n",
    "                \n",
    "            else:\n",
    "                first = mid +1\n",
    "                \n",
    "    return index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tryextract(i, pattern):\n",
    "    \n",
    "    try: \n",
    "        m = re.search(pattern, i).group(1)\n",
    "        return(m)\n",
    "    \n",
    "    except AttributeError:\n",
    "        return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with gzip.open(gff_file_path, \"rt\", encoding=\"utf-8\") as z:\n",
    "    \n",
    "    df = pd.read_csv(z,delimiter=r\"\\s+\",skiprows=1,header=None)\n",
    "    df.columns = ['chromosome', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    df['Name'] = df.attributes.apply(lambda x: tryextract(x, r\"Name=(.+);pacid\"))\n",
    "    #df['pacid'] = df.attributes.apply(lambda x: tryextract(x, r\"pacid=(.+);longest\"))\n",
    "    df['ID'] = df.attributes.apply(lambda x: tryextract(x, r\"ID=(PAC:[0-9]+)\"))\n",
    "        #except AttributeError: \n",
    "    #del df['attributes']\n",
    "    #df = df.sort_values(by=\"Name\")\n",
    "    df = df[df.feature == 'mRNA']\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset = df[df.feature ==\"mRNA\"].sort_values(by=\"Name\").reset_index()\n",
    "search_list = list(subset.Name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken 11.651297092437744 s\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "pacid_list = []\n",
    "\n",
    "for row in range(0,len(merged)):\n",
    "    transcript_id = merged.loc[row,'transcript_id_v5.3.1']\n",
    "    pacid = None\n",
    "    \n",
    "    index = BinarySearch(search_list, transcript_id)\n",
    "    \n",
    "    if index > -1: \n",
    "        pacid = subset.loc[index, \"ID\"]\n",
    "                    \n",
    "    pacid_list.append(pacid)\n",
    "\n",
    "t1 = time.time()\n",
    "print(\"Time taken\", t1-t0, \"s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged['PAC_ID'] = pacid_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define candidate gene conditions\n",
    "A = (merged.num_detected>=2)\n",
    "B = (merged.num_sampled>=2)\n",
    "C = merged.pathway_id.str.contains(\"TRIGLSYN-PWY\")==True\n",
    "D = merged.pathway_id.str.contains(\"PWY-4381\")==True\n",
    "F = (merged.num_detected>=1)\n",
    "\n",
    "condition = (A & (C | D))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set= []\n",
    "filtered_df = df.dropna()\n",
    "for transcript in list(merged[condition].PAC_ID):\n",
    "    temp = filtered_df[filtered_df.ID==transcript]\n",
    "    for i in list(temp.index):\n",
    "        #tp = (temp.chromosome[i], temp.start[i], temp.end[i])\n",
    "        gene_set.append((temp.chromosome[i], temp.start[i], temp.end[i]))\n",
    "    #if count>10: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/intermediate_data_from_gff/gene_set.pickle', 'wb') as f:\n",
    "    pickle.dump(gene_set, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13446    chromosome_1\n",
       "Name: chromosome, dtype: object"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "176 transcripts out of 17956 transcripts did not have a matching PAC ID\n"
     ]
    }
   ],
   "source": [
    "print(list(merged.PAC_ID).count(None),\"transcripts out of\", len(merged), \"transcripts did not have a matching PAC ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>gene</td>\n",
       "      <td>24026</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Cre01.g000050;Name=Cre01.g000050</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24026</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339;Name=Cre01.g000050.t1.3;pacid=...</td>\n",
       "      <td>Cre01.g000050.t1.3</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>24026</td>\n",
       "      <td>24125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339.five_prime_UTR.1;Parent=PAC:26...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>24126</td>\n",
       "      <td>28105</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=PAC:26903339.CDS.1;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28291</td>\n",
       "      <td>28644</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=PAC:26903339.CDS.2;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28842</td>\n",
       "      <td>29091</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=PAC:26903339.CDS.3;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>29347</td>\n",
       "      <td>29577</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=PAC:26903339.CDS.4;Parent=PAC:26903339;paci...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>29578</td>\n",
       "      <td>30617</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903339.three_prime_UTR.1;Parent=PAC:2...</td>\n",
       "      <td>None</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>gene</td>\n",
       "      <td>30776</td>\n",
       "      <td>41037</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Cre01.g000100;Name=Cre01.g000100</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chromosome_1</td>\n",
       "      <td>phytozome8_0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>30776</td>\n",
       "      <td>41037</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PAC:26903974;Name=Cre01.g000100.t1.3;pacid=...</td>\n",
       "      <td>Cre01.g000100.t1.3</td>\n",
       "      <td>PAC:26903974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome        source          feature  start    end score strand  \\\n",
       "0  chromosome_1  phytozome8_0             gene  24026  30617     .      +   \n",
       "1  chromosome_1  phytozome8_0             mRNA  24026  30617     .      +   \n",
       "2  chromosome_1  phytozome8_0   five_prime_UTR  24026  24125     .      +   \n",
       "3  chromosome_1  phytozome8_0              CDS  24126  28105     .      +   \n",
       "4  chromosome_1  phytozome8_0              CDS  28291  28644     .      +   \n",
       "5  chromosome_1  phytozome8_0              CDS  28842  29091     .      +   \n",
       "6  chromosome_1  phytozome8_0              CDS  29347  29577     .      +   \n",
       "7  chromosome_1  phytozome8_0  three_prime_UTR  29578  30617     .      +   \n",
       "8  chromosome_1  phytozome8_0             gene  30776  41037     .      +   \n",
       "9  chromosome_1  phytozome8_0             mRNA  30776  41037     .      +   \n",
       "\n",
       "  phase                                         attributes  \\\n",
       "0     .                ID=Cre01.g000050;Name=Cre01.g000050   \n",
       "1     .  ID=PAC:26903339;Name=Cre01.g000050.t1.3;pacid=...   \n",
       "2     .  ID=PAC:26903339.five_prime_UTR.1;Parent=PAC:26...   \n",
       "3     0  ID=PAC:26903339.CDS.1;Parent=PAC:26903339;paci...   \n",
       "4     1  ID=PAC:26903339.CDS.2;Parent=PAC:26903339;paci...   \n",
       "5     1  ID=PAC:26903339.CDS.3;Parent=PAC:26903339;paci...   \n",
       "6     0  ID=PAC:26903339.CDS.4;Parent=PAC:26903339;paci...   \n",
       "7     .  ID=PAC:26903339.three_prime_UTR.1;Parent=PAC:2...   \n",
       "8     .                ID=Cre01.g000100;Name=Cre01.g000100   \n",
       "9     .  ID=PAC:26903974;Name=Cre01.g000100.t1.3;pacid=...   \n",
       "\n",
       "                 Name            ID  \n",
       "0                None          None  \n",
       "1  Cre01.g000050.t1.3  PAC:26903339  \n",
       "2                None  PAC:26903339  \n",
       "3                None  PAC:26903339  \n",
       "4                None  PAC:26903339  \n",
       "5                None  PAC:26903339  \n",
       "6                None  PAC:26903339  \n",
       "7                None  PAC:26903339  \n",
       "8                None          None  \n",
       "9  Cre01.g000100.t1.3  PAC:26903974  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>num_detected</th>\n",
       "      <th>num_sampled</th>\n",
       "      <th>proportion</th>\n",
       "      <th>source</th>\n",
       "      <th>annotation_version</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>transcript_id_v5.3.1</th>\n",
       "      <th>PAC_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cre01.g000017.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016', 'Kwak_2017']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g2.t1</td>\n",
       "      <td>PAC:26903746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre01.g000033.t1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g3.t1</td>\n",
       "      <td>PAC:26903463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre01.g000050.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Gargouri_2015']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000050.t1.3</td>\n",
       "      <td>PAC:26903339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cre01.g000100.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000100.t1.3</td>\n",
       "      <td>PAC:26903974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cre01.g000150.t1.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000150</td>\n",
       "      <td>ZRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000150.t1.2</td>\n",
       "      <td>PAC:26903809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cre01.g000200.t1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000200.t1.3</td>\n",
       "      <td>PAC:26903011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cre01.g000250.t1.2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000250.t1.2</td>\n",
       "      <td>PAC:26903028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Cre01.g000300.t1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Creinhardtii PWY-5667', 'Creinhardtii PWY-74...</td>\n",
       "      <td>Cre01.g000300.t1.3</td>\n",
       "      <td>PAC:26903629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cre01.g000350.t1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.125</td>\n",
       "      <td>['Hemme_2014', 'Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000350.t1.3</td>\n",
       "      <td>PAC:26903061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Cre01.g000400.t1.2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>['Bajhaiya_2016']</td>\n",
       "      <td>v5.5</td>\n",
       "      <td>Cre01.g000400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre01.g000400.t1.2</td>\n",
       "      <td>PAC:26903690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id  num_detected  num_sampled  proportion  \\\n",
       "0  Cre01.g000017.t1.1             0            2       0.000   \n",
       "1  Cre01.g000033.t1.1             1            1       0.125   \n",
       "2  Cre01.g000050.t1.1             0            1       0.000   \n",
       "3  Cre01.g000100.t1.1             0            1       0.000   \n",
       "4  Cre01.g000150.t1.2             1            1       0.125   \n",
       "5  Cre01.g000200.t1.1             0            1       0.000   \n",
       "6  Cre01.g000250.t1.2             0            1       0.000   \n",
       "7  Cre01.g000300.t1.1             1            1       0.125   \n",
       "8  Cre01.g000350.t1.1             1            2       0.125   \n",
       "9  Cre01.g000400.t1.2             0            1       0.000   \n",
       "\n",
       "                            source annotation_version        gene_id  \\\n",
       "0   ['Bajhaiya_2016', 'Kwak_2017']               v5.5  Cre01.g000017   \n",
       "1                ['Bajhaiya_2016']               v5.5  Cre01.g000033   \n",
       "2                ['Gargouri_2015']               v5.5  Cre01.g000050   \n",
       "3                ['Bajhaiya_2016']               v5.5  Cre01.g000100   \n",
       "4                ['Bajhaiya_2016']               v5.5  Cre01.g000150   \n",
       "5                ['Bajhaiya_2016']               v5.5  Cre01.g000200   \n",
       "6                ['Bajhaiya_2016']               v5.5  Cre01.g000250   \n",
       "7                ['Bajhaiya_2016']               v5.5  Cre01.g000300   \n",
       "8  ['Hemme_2014', 'Bajhaiya_2016']               v5.5  Cre01.g000350   \n",
       "9                ['Bajhaiya_2016']               v5.5  Cre01.g000400   \n",
       "\n",
       "  gene_symbol                                         pathway_id  \\\n",
       "0         NaN                                                NaN   \n",
       "1         NaN                                                NaN   \n",
       "2         NaN                                                NaN   \n",
       "3         NaN                                                NaN   \n",
       "4        ZRT2                                                NaN   \n",
       "5         NaN                                                NaN   \n",
       "6         NaN                                                NaN   \n",
       "7         NaN  ['Creinhardtii PWY-5667', 'Creinhardtii PWY-74...   \n",
       "8         NaN                                                NaN   \n",
       "9         NaN                                                NaN   \n",
       "\n",
       "  transcript_id_v5.3.1        PAC_ID  \n",
       "0                g2.t1  PAC:26903746  \n",
       "1                g3.t1  PAC:26903463  \n",
       "2   Cre01.g000050.t1.3  PAC:26903339  \n",
       "3   Cre01.g000100.t1.3  PAC:26903974  \n",
       "4   Cre01.g000150.t1.2  PAC:26903809  \n",
       "5   Cre01.g000200.t1.3  PAC:26903011  \n",
       "6   Cre01.g000250.t1.2  PAC:26903028  \n",
       "7   Cre01.g000300.t1.3  PAC:26903629  \n",
       "8   Cre01.g000350.t1.3  PAC:26903061  \n",
       "9   Cre01.g000400.t1.2  PAC:26903690  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'chromosome'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-617a9e7c64d6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;31m#print(key, value)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgff_ext_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m         \u001b[0mtp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m         \u001b[0mgene_set\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcount\u001b[0m \u001b[0;34m>\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'chromosome'"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "gene_set = []\n",
    "for key in gff_ext_dict.keys():\n",
    "    #print(key, value)\n",
    "    for row in gff_ext_dict[key]:\n",
    "        tp = (row.chromosome, row.start, row.end)\n",
    "        gene_set.append(tp)\n",
    "    if count >10: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
