{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate clean_primary_data.pk\n",
    "**Output:** `../../data/workflow/clean_primary_data.pk`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle as pk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-3241ea60d731>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprimary_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_excel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"../../data/raw_data/Summary_of_primary_data.xlsx\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Sheet1'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprimary_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/alice_personal/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    297\u001b[0m                 )\n\u001b[1;32m    298\u001b[0m                 \u001b[0mwarnings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mFutureWarning\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 299\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    300\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/alice_personal/lib/python3.9/site-packages/pandas/io/excel/_base.py\u001b[0m in \u001b[0;36mread_excel\u001b[0;34m(io, sheet_name, header, names, index_col, usecols, squeeze, dtype, engine, converters, true_values, false_values, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, parse_dates, date_parser, thousands, comment, skipfooter, convert_float, mangle_dupe_cols, storage_options)\u001b[0m\n\u001b[1;32m    334\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mExcelFile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m         \u001b[0mshould_close\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 336\u001b[0;31m         \u001b[0mio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mExcelFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    337\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m         raise ValueError(\n",
      "\u001b[0;32m~/.conda/envs/alice_personal/lib/python3.9/site-packages/pandas/io/excel/_base.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, path_or_buffer, engine, storage_options)\u001b[0m\n\u001b[1;32m   1129\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstorage_options\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstorage_options\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1130\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1131\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engines\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_io\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1132\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__fspath__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/alice_personal/lib/python3.9/site-packages/pandas/io/excel/_xlrd.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filepath_or_buffer, storage_options)\u001b[0m\n\u001b[1;32m     22\u001b[0m         \"\"\"\n\u001b[1;32m     23\u001b[0m         \u001b[0merr_msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"Install xlrd >= 1.0.0 for Excel support\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m         \u001b[0mimport_optional_dependency\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"xlrd\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextra\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstorage_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/alice_personal/lib/python3.9/site-packages/pandas/compat/_optional.py\u001b[0m in \u001b[0;36mimport_optional_dependency\u001b[0;34m(name, extra, raise_on_missing, on_version)\u001b[0m\n\u001b[1;32m    107\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mraise_on_missing\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 109\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    110\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    111\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd."
     ]
    }
   ],
   "source": [
    "primary_data = pd.read_excel(\"../../data/raw_data/Summary_of_primary_data.xlsx\", 'Sheet1')\n",
    "primary_data[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>First_author</th>\n",
       "      <th>Publication_year</th>\n",
       "      <th>Type_of_study</th>\n",
       "      <th>Genes_sampled</th>\n",
       "      <th>Genes_manipulated</th>\n",
       "      <th>Genes_with_effect</th>\n",
       "      <th>Genes_no_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Chlamydomonas</td>\n",
       "      <td>reinhardtii</td>\n",
       "      <td>Lv</td>\n",
       "      <td>2013</td>\n",
       "      <td>knock_out; transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cre05.g248150.t1; Cre06.g299050.t1.2</td>\n",
       "      <td>Cre06.g299050.t1.2</td>\n",
       "      <td>Cre05.g248150.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>Chlamydomonas</td>\n",
       "      <td>reinhardtii</td>\n",
       "      <td>Miller</td>\n",
       "      <td>2010</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>Chlamydomonas</td>\n",
       "      <td>reinhardtii</td>\n",
       "      <td>Kwak</td>\n",
       "      <td>2017</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index          Genus      Species First_author  Publication_year  \\\n",
       "0      1  Chlamydomonas  reinhardtii           Lv              2013   \n",
       "1      5  Chlamydomonas  reinhardtii       Miller              2010   \n",
       "2      7  Chlamydomonas  reinhardtii         Kwak              2017   \n",
       "\n",
       "                Type_of_study Genes_sampled  \\\n",
       "0  knock_out; transcriptomics           NaN   \n",
       "1             transcriptomics           NaN   \n",
       "2             transcriptomics           NaN   \n",
       "\n",
       "                      Genes_manipulated   Genes_with_effect   Genes_no_effect  \n",
       "0  Cre05.g248150.t1; Cre06.g299050.t1.2  Cre06.g299050.t1.2  Cre05.g248150.t1  \n",
       "1                                   NaN                 NaN               NaN  \n",
       "2                                   NaN                 NaN               NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primary_data = primary_data[['Genus', 'Species','First_author','Publication_year',\\\n",
    "                             'Type_of_study','Genes_sampled','Genes_manipulated', 'Genes_with_effect',\\\n",
    "                             \"Genes_no_effect\"]]\n",
    "subset = primary_data[primary_data.Genus=='Chlamydomonas'][primary_data.Species=='reinhardtii'].reset_index()\n",
    "subset[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "subset.Publication_year = subset.Publication_year.astype(str)\n",
    "df['source'] = subset.First_author+\"_\"+subset.Publication_year\n",
    "df['manipulated_gene'] = subset.apply(lambda x: 1 if x.Genes_manipulated is not np.nan else 0, axis =1)\n",
    "df['type_of_study']  = subset.apply(lambda x: x.Type_of_study.split(\"; \"), axis =1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>manipulated_gene</th>\n",
       "      <th>type_of_study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out, transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Miller_2010</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kwak_2017</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        source  manipulated_gene                 type_of_study\n",
       "0      Lv_2013                 1  [knock_out, transcriptomics]\n",
       "1  Miller_2010                 0             [transcriptomics]\n",
       "2    Kwak_2017                 0             [transcriptomics]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/workflow/clean_primary_data.pk', 'wb') as f:\n",
    "    pk.dump(df, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['RT_qPCR', 'gene_silence', 'genetic_transformation', 'genomics', 'knock_out', 'proteomics', 'transcriptomics']\n"
     ]
    }
   ],
   "source": [
    "types_of_study = list(np.unique([i for sublist in df.type_of_study for i in sublist]))\n",
    "print(types_of_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../data/workflow/clean_primary_data.pk', 'rb') as f:\n",
    "    df = pd.read_pickle(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>manipulated_gene</th>\n",
       "      <th>type_of_study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bajhaiya_2016</td>\n",
       "      <td>1</td>\n",
       "      <td>[transcriptomics, RT_qPCR, knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blaby_2013</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out, transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Boyle_2012a</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Boyle_2012b</td>\n",
       "      <td>1</td>\n",
       "      <td>[RT_qPCR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Couso_2016</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Deng_2011</td>\n",
       "      <td>1</td>\n",
       "      <td>[gene_silence]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Deng_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[gene_silence]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fang_2012</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Gargouri_2015</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics, proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Goodenough_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Goodson_2011</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Hemme_2014</td>\n",
       "      <td>0</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Iwai_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[genetic_transformation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>James_2011</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>James_2013</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Juergens_2015</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics, proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Kao_2017</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR, genetic_transformation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Kwak_2017</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>La Russa_2012</td>\n",
       "      <td>1</td>\n",
       "      <td>[genetic_transformation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Lee_2012</td>\n",
       "      <td>0</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Legret_2015</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR, transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Li_2010</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Li_2012a</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Li_2012b</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Li_2016</td>\n",
       "      <td>0</td>\n",
       "      <td>[genomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Longworth_2012</td>\n",
       "      <td>0</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Lv_2013</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out, transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Miller_2010</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Moellering_2009a</td>\n",
       "      <td>1</td>\n",
       "      <td>[gene_silence]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Moellering_2009b</td>\n",
       "      <td>0</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Msanne_2012</td>\n",
       "      <td>1</td>\n",
       "      <td>[RT_qPCR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Nguyen_2011</td>\n",
       "      <td>1</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Nguyen_2013</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR, genetic_transformation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Park_2015</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics, proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Ramanan_2013</td>\n",
       "      <td>1</td>\n",
       "      <td>[RT_qPCR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Riekhof_2003</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Sato_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[RT_qPCR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Schulz-Raffelt_2016</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Shtaida_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[gene_silence]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Siaut_2011</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Urzica_2013</td>\n",
       "      <td>0</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Valledor_2013</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR, proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Valledor_2014</td>\n",
       "      <td>0</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Velmurugan_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>[proteomics]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Wang_2009</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Work_2010</td>\n",
       "      <td>1</td>\n",
       "      <td>[knock_out]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Yamaoka_2016</td>\n",
       "      <td>0</td>\n",
       "      <td>[RT_qPCR, genetic_transformation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Yoon_2012</td>\n",
       "      <td>1</td>\n",
       "      <td>[gene_silence]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 source  manipulated_gene  \\\n",
       "0         Bajhaiya_2016                 1   \n",
       "1            Blaby_2013                 1   \n",
       "2           Boyle_2012a                 1   \n",
       "3           Boyle_2012b                 1   \n",
       "4            Couso_2016                 1   \n",
       "5             Deng_2011                 1   \n",
       "6             Deng_2014                 1   \n",
       "7             Fang_2012                 1   \n",
       "8         Gargouri_2015                 0   \n",
       "9       Goodenough_2014                 1   \n",
       "10         Goodson_2011                 1   \n",
       "11           Hemme_2014                 0   \n",
       "12            Iwai_2014                 1   \n",
       "13           James_2011                 1   \n",
       "14           James_2013                 1   \n",
       "15        Juergens_2015                 0   \n",
       "16             Kao_2017                 0   \n",
       "17            Kwak_2017                 0   \n",
       "18        La Russa_2012                 1   \n",
       "19             Lee_2012                 0   \n",
       "20          Legret_2015                 0   \n",
       "21              Li_2010                 1   \n",
       "22             Li_2012a                 1   \n",
       "23             Li_2012b                 0   \n",
       "24              Li_2016                 0   \n",
       "25       Longworth_2012                 0   \n",
       "26              Lv_2013                 1   \n",
       "27          Miller_2010                 0   \n",
       "28     Moellering_2009a                 1   \n",
       "29     Moellering_2009b                 0   \n",
       "30          Msanne_2012                 1   \n",
       "31          Nguyen_2011                 1   \n",
       "32          Nguyen_2013                 0   \n",
       "33            Park_2015                 0   \n",
       "34         Ramanan_2013                 1   \n",
       "35         Riekhof_2003                 1   \n",
       "36            Sato_2014                 1   \n",
       "37  Schulz-Raffelt_2016                 1   \n",
       "38         Shtaida_2014                 1   \n",
       "39           Siaut_2011                 1   \n",
       "40          Urzica_2013                 0   \n",
       "41        Valledor_2013                 0   \n",
       "42        Valledor_2014                 0   \n",
       "43      Velmurugan_2014                 1   \n",
       "44            Wang_2009                 1   \n",
       "45            Work_2010                 1   \n",
       "46         Yamaoka_2016                 0   \n",
       "47            Yoon_2012                 1   \n",
       "\n",
       "                            type_of_study  \n",
       "0   [transcriptomics, RT_qPCR, knock_out]  \n",
       "1            [knock_out, transcriptomics]  \n",
       "2                             [knock_out]  \n",
       "3                               [RT_qPCR]  \n",
       "4                             [knock_out]  \n",
       "5                          [gene_silence]  \n",
       "6                          [gene_silence]  \n",
       "7                             [knock_out]  \n",
       "8           [transcriptomics, proteomics]  \n",
       "9                       [transcriptomics]  \n",
       "10                            [knock_out]  \n",
       "11                           [proteomics]  \n",
       "12               [genetic_transformation]  \n",
       "13                            [knock_out]  \n",
       "14                            [knock_out]  \n",
       "15          [transcriptomics, proteomics]  \n",
       "16      [RT_qPCR, genetic_transformation]  \n",
       "17                      [transcriptomics]  \n",
       "18               [genetic_transformation]  \n",
       "19                           [proteomics]  \n",
       "20             [RT_qPCR, transcriptomics]  \n",
       "21                            [knock_out]  \n",
       "22                            [knock_out]  \n",
       "23                              [RT_qPCR]  \n",
       "24                             [genomics]  \n",
       "25                           [proteomics]  \n",
       "26           [knock_out, transcriptomics]  \n",
       "27                      [transcriptomics]  \n",
       "28                         [gene_silence]  \n",
       "29                           [proteomics]  \n",
       "30                              [RT_qPCR]  \n",
       "31                           [proteomics]  \n",
       "32      [RT_qPCR, genetic_transformation]  \n",
       "33          [transcriptomics, proteomics]  \n",
       "34                              [RT_qPCR]  \n",
       "35                            [knock_out]  \n",
       "36                              [RT_qPCR]  \n",
       "37                            [knock_out]  \n",
       "38                         [gene_silence]  \n",
       "39                            [knock_out]  \n",
       "40                      [transcriptomics]  \n",
       "41                  [RT_qPCR, proteomics]  \n",
       "42                           [proteomics]  \n",
       "43                           [proteomics]  \n",
       "44                            [knock_out]  \n",
       "45                            [knock_out]  \n",
       "46      [RT_qPCR, genetic_transformation]  \n",
       "47                         [gene_silence]  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
